<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.515">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Samuele Soraggi">
<meta name="dcterms.date" content="2024-10-03">
<meta name="description" content="Learn how to download (especially) GEO-SRA data">

<title>ABC - Accessible Bioinformatics Cafe - Download from online databases, open coding session</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-WG01Z2LJ1J"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-WG01Z2LJ1J', { 'anonymize_ip': true});
</script>
<!-- MailerLite Universal -->
<script>
    (function(w,d,e,u,f,l,n){w[f]=w[f]||function(){(w[f].q=w[f].q||[])
    .push(arguments);},l=d.createElement(e),l.async=1,l.src=u,
    n=d.getElementsByTagName(e)[0],n.parentNode.insertBefore(l,n);})
    (window,document,'script','https://assets.mailerlite.com/js/universal.js','ml');
    ml('account', '1004787');
</script>
<!-- End MailerLite Universal -->


<meta property="og:title" content="ABC - Accessible Bioinformatics Cafe - Download from online databases, open coding session">
<meta property="og:description" content="Learn how to download (especially) GEO-SRA data">
<meta property="og:image" content="https://abc.au.dk
litesrv._domainkey.birc.au.dk/documentation/2024-10-03-ABC6/dna.jpg">
<meta property="og:site_name" content="ABC - Accessible Bioinformatics Cafe">
<meta property="og:locale" content="en_US">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../img/ABClogo.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ABC - Accessible Bioinformatics Cafe</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../news.html"> 
<span class="menu-text">Calendar</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Documentation.html"> 
<span class="menu-text">Documentation</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-resources" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Resources</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-resources">    
        <li>
    <a class="dropdown-item" href="../binfResources/bash.html">
 <span class="dropdown-text">Bash tips</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../gethelp.html"> 
<span class="menu-text">People&amp;Help</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/AU-ABC"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#slides" id="toc-slides" class="nav-link active" data-scroll-target="#slides">Slides</a></li>
  <li><a href="#download-from-geo-and-sra" id="toc-download-from-geo-and-sra" class="nav-link" data-scroll-target="#download-from-geo-and-sra">Download from GEO and SRA</a>
  <ul class="collapse">
  <li><a href="#geo-and-sra-data" id="toc-geo-and-sra-data" class="nav-link" data-scroll-target="#geo-and-sra-data">GEO and SRA data</a>
  <ul class="collapse">
  <li><a href="#download-geo-data-programmatically" id="toc-download-geo-data-programmatically" class="nav-link" data-scroll-target="#download-geo-data-programmatically">Download GEO data programmatically</a></li>
  <li><a href="#download-geo-and-sra-data-with-geofetch" id="toc-download-geo-and-sra-data-with-geofetch" class="nav-link" data-scroll-target="#download-geo-and-sra-data-with-geofetch">Download GEO and SRA data with <code>geofetch</code></a></li>
  <li><a href="#download-sra-data-with-fasterq-dump" id="toc-download-sra-data-with-fasterq-dump" class="nav-link" data-scroll-target="#download-sra-data-with-fasterq-dump">Download SRA data with <code>fasterq-dump</code></a></li>
  </ul></li>
  </ul></li>
  <li><a href="#extra-references-for-fastq-dump-and-fasterq-dump" id="toc-extra-references-for-fastq-dump-and-fasterq-dump" class="nav-link" data-scroll-target="#extra-references-for-fastq-dump-and-fasterq-dump">Extra references for <code>fastq-dump</code> and <code>fasterq-dump</code></a>
  <ul class="collapse">
  <li><a href="#fastq-dump-command" id="toc-fastq-dump-command" class="nav-link" data-scroll-target="#fastq-dump-command">fastq-dump command</a>
  <ul class="collapse">
  <li><a href="#downloading-reads-from-the-sra-using-fastq-dump" id="toc-downloading-reads-from-the-sra-using-fastq-dump" class="nav-link" data-scroll-target="#downloading-reads-from-the-sra-using-fastq-dump">Downloading reads from the SRA using <code>fastq-dump</code></a></li>
  </ul></li>
  <li><a href="#fasterq-dump-command" id="toc-fasterq-dump-command" class="nav-link" data-scroll-target="#fasterq-dump-command">fasterq-dump command</a>
  <ul class="collapse">
  <li><a href="#fasterq-dump-vs-fastq-dump" id="toc-fasterq-dump-vs-fastq-dump" class="nav-link" data-scroll-target="#fasterq-dump-vs-fastq-dump"><code>fasterq-dump</code> vs <code>fastq-dump</code></a></li>
  <li><a href="#downloading-reads-from-the-sra-using-fasterq-dump" id="toc-downloading-reads-from-the-sra-using-fasterq-dump" class="nav-link" data-scroll-target="#downloading-reads-from-the-sra-using-fasterq-dump">Downloading reads from the SRA using <code>fasterq-dump</code></a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Download from online databases, open coding session</h1>
  <div class="quarto-categories">
    <div class="quarto-category">GEO</div>
    <div class="quarto-category">SRA</div>
    <div class="quarto-category">Download</div>
  </div>
  </div>

<div>
  <div class="description">
    Learn how to download (especially) GEO-SRA data
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Samuele Soraggi </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">October 3, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="slides" class="level1">
<h1>Slides</h1>
<p>Today’s slides</p>
<p>&nbsp;</p>
<p align="center">
</p><p><a href="https://abc.au.dk/documentation/slides/20241003-ABC6.zip" style="background-color: #4266A1; color: #FFFFFF; padding: 30px 20px; text-decoration: none; border-radius: 5px;"> Download Slides </a></p>
<p></p>
<p>&nbsp;</p>
</section>
<section id="download-from-geo-and-sra" class="level1">
<h1>Download from GEO and SRA</h1>
<p>There is a multitude of online databases for OMICS data of all types (we are trying to put together <a href="../binfResources/Databases.html">an useful list of databases here</a>). Some of those databases provide direct links to the data to download (either raw data or analysis results), some others havea more complex way of downloading data (through a software or some more complex way of finding the download link). A database people often find problematic is <a href="https://www.ncbi.nlm.nih.gov/geo/">GEO</a> (Gene Expression Omnibus), which includes processed and raw data, usally for short read datasets.</p>
<section id="geo-and-sra-data" class="level2">
<h2 class="anchored" data-anchor-id="geo-and-sra-data">GEO and SRA data</h2>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/">GEO</a> (Gene Expression Omnibus) is a database which contains processed data from biological experiments. For example aligned files and analysis results, and supplementary files of various type. A GEO entry is called <strong>Series</strong> and contains <strong>Samples</strong> , <strong>Platforms</strong>, <strong>Supplementary data</strong>. Usually, there is an associate SRA number.</p>
<p>A <a href="https://www.ncbi.nlm.nih.gov/sra/">SRA</a> (Sequence Read Archive) number contains the <strong>raw data for the samples contained in a GEO series</strong>. The download from GEO can happen by using the various alphanumerical identifiers to write a download link. SRA downloads need the software SRA-tools, and cannot be performed using a simple download link.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./2024-10-03-ABC6/GEOpage.png" class="img-fluid figure-img"></p>
<figcaption>A GEO Series, where highlighted in different colours there are the main data elements. The Series ID (red). the type of platform (green) for creating the samples (blue), and the SRA number for raw data (purple).</figcaption>
</figure>
</div>
<section id="download-geo-data-programmatically" class="level3">
<h3 class="anchored" data-anchor-id="download-geo-data-programmatically">Download GEO data programmatically</h3>
<p>Sometimes you need to download some supplementary files from GEO. This is pretty easy from the command line, and we show it with an example. We consider the GEO series in the figure above - go to the <a href="https://www.ncbi.nlm.nih.gov/geo/">GEO home page</a> and find it using the series number in the search bar of the website.</p>
<p>Now, we use the <a href="https://www.ncbi.nlm.nih.gov/geo/info/geo_paccess.html">guidelines of GEO for programmatic downloads</a>. Here you have examples and descriptions on how to write the FTP<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> address of all files in the series (excluding SRA files).</p>
<p>To look at the content of the series, use the command line and write</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-l</span> ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE67nnn/GSE67303/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>note how the address contains <code>series</code>, the series ID where the last three numbers are changed into <code>nnn</code>, and the series number itself. The output is a list of the content of the series:</p>
<pre><code>matrix
miniml
soft
suppl</code></pre>
<p>The first three are folders containing matrix, miniml and soft files, which are plain text files used to describe the series. The fourth is the folder of supplementary material. Using <code>curl -l</code> we can see that it contains the supplementary file of the series:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-l</span> ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE67nnn/GSE67303/suppl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>GSE67303_DEG_cuffdiff.xlsx</code></pre>
<p>Now we can download the file using the whole URL with <code>wget</code></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE67nnn/GSE67303/suppl/GSE67303_DEG_cuffdiff.xlsx</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Do you want to download a lot of supplementary files without one-by-one download by hand? Then <code>wget -r</code> will download recursively the content of the whole folder, for example:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-r</span> ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE67nnn/GSE67303/suppl/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Some webpages have so-called robots which avoid a user to download in an autopmated way too many datasets at once, to prevent overloading of the bandwidth. Bulk downloads are identified by, amongst others, how much time it takes to request a download. When a robot rejects your download attempt, you get an error message. In that case, <code>wget</code> can solve the problems with some extra dedicated options. Below you can see a bulk download from the <a href="https://www.ebi.ac.uk/pride/">PRIDE</a> database, which uses robots:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">--random-wait</span> <span class="dt">\</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">-r</span> <span class="at">-p</span> <span class="at">-e</span> robots=off <span class="dt">\</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">-U</span> mozilla <span class="dt">\</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>     ftp.pride.ebi.ac.uk/pride/data/archive/2024/09/PXD056312/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Note: the command above takes time because the data is quite big and is just for the sake of example, so don’t run it :-)</p>
</div>
</div>
<p>You can change the URL to show the supplementary data inside samples and inside platforms. In our example you can click on the platform/sample number on the webpage, and see if they contain any supplementary data. It is not the case of this GEO series, but for the sake of the example we will write down how to list the supplementary files inside the related FTP folders.</p>
<ul>
<li><p>For the platforms, we need to substute <code>series</code> with <code>platforms</code>, and the same with the IDs, which will again have <code>nnn</code> for the last 3 digits: <code>curl -l ftp://ftp.ncbi.nlm.nih.gov/geo/platforms/GPL19nnn/GPL19949/</code> you will notice there are only <code>soft</code> and <code>miniml</code> folders for the platform description, and no supplementary</p></li>
<li><p>For the samples, it works analogously, using sample IDs <code>curl -l ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1644nnn/GSM1644066/</code> you will notice there are only <code>soft</code> and <code>miniml</code> folders for the platform description</p></li>
</ul>
</section>
<section id="download-geo-and-sra-data-with-geofetch" class="level3">
<h3 class="anchored" data-anchor-id="download-geo-and-sra-data-with-geofetch">Download GEO and SRA data with <code>geofetch</code></h3>
<p>Downloading entire GEO series and related raw data through SRA is always painful. Here comes into play a pretty useful package called <code>geofetch</code> (<span class="citation" data-cites="khoroshevskyi_geofetch:_2023">Khoroshevskyi et al. (<a href="#ref-khoroshevskyi_geofetch:_2023" role="doc-biblioref">2023</a>)</span>). This can be easily installed through <code>conda</code> on the command line</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-n</span> geofetch</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate geofetch</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> bioconda <span class="at">-c</span> conda-forge geofetch sra-tools</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout-tips">
<p>You can also find a guide to <code>anaconda</code> in <a href="../documentation/2024-06-27-ABC2.html">our ABC2 tutorial</a>, where the packages <code>geofetch</code> and <code>sra-tools</code> can be installed interactively. <code>geofetch</code> works also with windows, though some options are not available.</p>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="settings for sra-tools">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
settings for sra-tools
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>sra-tools</code> creates very large files in <code>.sra</code> format into your home folder while downloading. This can be detrimental on a cluster, for example on GenomeDK, where your home is limited in 100GB size. <code>sra</code> files are then converted into the desired format. You can decide to download <code>sra</code> files in to a temporary folder or any other folder of your choice. For example you can do what follows to change the folder for <code>sra</code> files into <code>/tmp</code></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">${HOME}</span>/.ncbi</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'/repository/user/main/public/root = "/tmp"'</span> <span class="op">&gt;</span> <span class="va">${HOME}</span>/.ncbi/user-settings.mkfg</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<section id="download" class="level4">
<h4 class="anchored" data-anchor-id="download">Download</h4>
<p><code>geofetch</code> makes it easy to download data - the command is very basic and different options can personalize which content you want to download. The whole download contain also metadata which follows the PEP format, a FAIR-ready way of documenting data for reproducible access by other users (<span class="citation" data-cites="sheffield_linking_2021">Sheffield et al. (<a href="#ref-sheffield_linking_2021" role="doc-biblioref">2021</a>)</span>, <span class="citation" data-cites="leroy_pephub:_2024">LeRoy et al. (<a href="#ref-leroy_pephub:_2024" role="doc-biblioref">2024</a>)</span>). The basic command has this form:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">geofetch</span> <span class="at">-i</span> SRA_SERIES_ID</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>and one can add other arguments to choose the type of data to be downloaded according to this table:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./2024-10-03-ABC6/arguments_outputs.svg" class="img-fluid figure-img"></p>
<figcaption>Arguments to choose what to download with geofetch. The first line with empty arguments corresponds to the basic geofetch command.</figcaption>
</figure>
</div>
<p>We can download metadata and raw <code>sra</code> data for the GEO series we have in the first figure of this tutorial with the basic <code>geofetch</code> command</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">geofetch</span> <span class="at">-i</span> GSE67303</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Once you are done, you have the following files and folders:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">GSE67303</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>GSE67303
├── GSE67303_GSE.soft
├── GSE67303_GSM.soft
├── GSE67303_PEP
│&nbsp;&nbsp; ├── GSE67303_PEP_raw.csv
│&nbsp;&nbsp; └── GSE67303_PEP.yaml
├── GSE67303_SRA.csv</code></pre>
<ul>
<li><code>GSE67303_GSE.soft</code> is a text file containing series information</li>
<li><code>GSE67303_GSM.soft</code> is a text file containing samples information</li>
<li><code>GSE67303_SRA.csv</code> is a comma-separated table with the <code>sra</code> file lists who was downloaded</li>
<li>the files in the subfolder <code>GSE67303_PEP</code> contains data info following <code>PEP</code> standard</li>
</ul>
<p>Try to do <code>ls</code> in the folder used as default for the download of <code>sra</code> data, such as</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> /tmp/sra/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>and you should see the downloaded data from the <code>sra</code> archive</p>
<pre><code>SRR1930183.sra  SRR1930184.sra  SRR1930185.sra  SRR1930186.sra</code></pre>
<p>Now you can use <code>fasterq-dump</code> to convert the <code>sra</code> data into <code>fastq</code> format. <code>fasterq-dump</code> builds up on the tool <code>fastq-dump</code>, but includes some default settings, such as splitting the <code>fastq</code> files into paired reads and ambiguous reads (only when you actually have paired reads), which is usually the standard.</p>
<p>In this example we choose the locally downloaded files in the <code>sra</code> download folder, and we choose the output folder inside where we have the metadata information.</p>
<pre><code>fasterq-dump /tmp/sra/*.sra -O GSE67303/raw -v</code></pre>
<p>Now you can see the raw data is also there</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> /tmp/sra/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>GSE67303
├── GSE67303_GSE.soft
├── GSE67303_GSM.soft
├── GSE67303_PEP
│&nbsp;&nbsp; ├── GSE67303_PEP_raw.csv
│&nbsp;&nbsp; └── GSE67303_PEP.yaml
├── GSE67303_SRA.csv
└── raw
    ├── SRR1930183_1.fastq
    ├── SRR1930183_2.fastq
    ├── SRR1930184_1.fastq
    ├── SRR1930184_2.fastq
    ├── SRR1930185_1.fastq
    ├── SRR1930185_2.fastq
    ├── SRR1930186_1.fastq
    └── SRR1930186_2.fastq</code></pre>
<p>If you want, you can empty the <code>sra</code> download folder</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> /tmp/sra/<span class="pp">*</span>.sra</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="download-sra-data-with-fasterq-dump" class="level3">
<h3 class="anchored" data-anchor-id="download-sra-data-with-fasterq-dump">Download SRA data with <code>fasterq-dump</code></h3>
<p>You can also use <code>fasterq-dump</code> directly if you do not want any metadata. It can be used with only one <code>sra</code> run at the time, but below we show how you can download entire series runs or sample runs.</p>
<section id="only-one-sra-run" class="level4">
<h4 class="anchored" data-anchor-id="only-one-sra-run">Only one SRA run</h4>
<p>Do you simply need a single raw file? In the webpage of the GEO series, click on the SRA code. A new window will show all the samples: choose one and find the SRA run number, which starts with SRR (in figure below).</p>
<p><img src="2024-10-03-ABC6/SRR.png" class="img-fluid"></p>
<p>In relation to figure above</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fasterq-dump</span> SRR1930186 <span class="at">-v</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="all-sra-runs-of-a-geo-series" class="level4">
<h4 class="anchored" data-anchor-id="all-sra-runs-of-a-geo-series">All SRA runs of a GEO series</h4>
<p>Click on the SRA link inside the GEO series. When the list with all samples appears, click on <code>Send to</code> on the top-right button, choose <code>File</code> and <code>Accession List</code>. This will download a <code>csv</code> file which you can combine with <code>fasterq-dump</code> to download all runs:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fasterq-dump</span> <span class="kw">`</span><span class="fu">sed</span> 1,1d ./Downloads/SraAccList.csv<span class="kw">`</span> <span class="at">-v</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="all-sra-runs-of-a-single-sample" class="level4">
<h4 class="anchored" data-anchor-id="all-sra-runs-of-a-single-sample">All SRA runs of a single sample</h4>
<p>After clicking on the SRA code of the GEO series, you can click on a specific sample and use the same method with the <code>Send to</code> button to download all the runs of a specific sample.</p>
<hr>
</section>
</section>
</section>
</section>
<section id="extra-references-for-fastq-dump-and-fasterq-dump" class="level1">
<h1>Extra references for <code>fastq-dump</code> and <code>fasterq-dump</code></h1>
<p>Below some references with all the options for getting <code>sra</code> raw data (credit <a href="https://github.com/rnnh">Ronan Harrington</a>)</p>
<section id="fastq-dump-command" class="level2">
<h2 class="anchored" data-anchor-id="fastq-dump-command">fastq-dump command</h2>
<p><code>fastq-dump</code> is a tool for downloading sequencing reads from <a href="https://www.ncbi.nlm.nih.gov/sra">NCBI’s Sequence Read Archive (SRA)</a>. These sequence reads will be downloaded as FASTQ files. How these FASTQ files are formatted depends on the <code>fastq-dump</code> options used.</p>
<section id="downloading-reads-from-the-sra-using-fastq-dump" class="level3">
<h3 class="anchored" data-anchor-id="downloading-reads-from-the-sra-using-fastq-dump">Downloading reads from the SRA using <code>fastq-dump</code></h3>
<p>In this example, we want to download FASTQ reads for a mate-pair library.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastq-dump</span> <span class="at">--gzip</span> <span class="at">--skip-technical</span> <span class="at">--readids</span> <span class="at">--read-filter</span> pass <span class="at">--dumpbase</span> <span class="at">--split-3</span> <span class="at">--clip</span> <span class="at">--outdir</span> path/to/reads/ SRR_ID</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>In this command…</p>
<ol type="1">
<li><strong><code>--gzip</code></strong>: Compress output using gzip. Gzip archived reads can be read directly by <code>bowtie2</code>.</li>
<li><strong><code>--skip-technical</code></strong>: Dump only biological reads, skip the technical reads.</li>
<li><strong><code>--readids</code></strong> or <strong><code>-I</code></strong>: Append read ID after spot ID as ‘accession.spot.readid’. With this flag, one sequence gets appended the ID <code>.1</code> and the other <code>.2</code>. Without this option, pair-ended reads will have identical IDs.</li>
<li><strong><code>--read-filter pass</code></strong>: Only returns reads that pass filtering (without <code>N</code>s).</li>
<li><strong><code>--dumpbase</code></strong> or <strong><code>-B</code></strong>: Formats sequence using base space (default for other than SOLiD). Included to avoid colourspace (in which pairs of bases are represented by numbers).</li>
<li><strong><code>--split-3</code></strong> separates the reads into left and right ends. If there is a left end without a matching right end, or a right end without a matching left end, they will be put in a single file.</li>
<li><strong><code>--clip</code></strong> or <strong><code>-W</code></strong>: Some of the sequences in the SRA contain tags that need to be removed. This will remove those sequences.</li>
<li><strong><code>--outdir</code></strong> or <strong><code>-O</code></strong>: <em>(Optional)</em> Output directory, default is current working directory.</li>
<li><strong><code>SRR_ID</code></strong>: This is is the ID of the run from SRA to be downloaded. This ID begins with “SRR” and is followed by around seven digits (e.g.&nbsp;<code>SRA1234567</code>).</li>
</ol>
<p>Other options that can be used instead of <code>--split-3</code>:</p>
<ol type="1">
<li><strong><code>--split-files</code></strong> splits the FASTQ reads into two files: one file for mate 1s (<code>...1</code>), and another for mate 2s (<code>..._2</code>). This option will not mateless pairs into a third file.</li>
<li><strong><code>--split-spot</code></strong> splits the FASTQ reads into two (mate 1s and mate 2s) within one file. <code>--split-spot</code> gives you an 8-line fastq format where forward precedes reverse (see <a href="https://www.biostars.org/p/178586/#258378" class="uri">https://www.biostars.org/p/178586/#258378</a>).</li>
</ol>
</section>
</section>
<section id="fasterq-dump-command" class="level2">
<h2 class="anchored" data-anchor-id="fasterq-dump-command">fasterq-dump command</h2>
<p><code>fasterq-dump</code> is a tool for downloading sequencing reads from <a href="https://www.ncbi.nlm.nih.gov/sra">NCBI’s Sequence Read Archive (SRA)</a>. These sequence reads will be downloaded as <code>fastq</code> files. <code>fasterq-dump</code> is a newer, streamlined alternative to <code>fastq-dump</code>; both of these programs are a part of <a href="https://anaconda.org/bioconda/sra-tools">sra-tools</a>.</p>
<section id="fasterq-dump-vs-fastq-dump" class="level3">
<h3 class="anchored" data-anchor-id="fasterq-dump-vs-fastq-dump"><code>fasterq-dump</code> vs <code>fastq-dump</code></h3>
<p>Here are a few of the differences between <code>fastq-dump</code> and <code>fasterq-dump</code>:</p>
<ol type="1">
<li>In <code>fastq-dump</code>, the flag <code>--split-3</code> is required to separate paired reads into left and right ends. This is the default setting in <code>fasterq-dump</code>.</li>
<li>The <code>fastq-dump</code> flag <code>--skip-technical</code> is no longer required to skip technical reads in <code>fasterq-dump</code>. Instead, the flag <code>--include-technical</code> is required to include technical reads when using <code>fasterq-dump</code>.</li>
<li>There is no <code>--gzip</code> or <code>--bzip2</code> flag in <code>fasterq-dump</code> to download compressed reads with <code>fasterq-dump</code>. However, FASTQ files downloaded using <code>fasterq-dump</code> can still be subsequently compressed.</li>
</ol>
<p>The following commands are equivalent, but will be executed faster using <code>fasterq-dump</code>:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastq-dump</span> SRR_ID <span class="at">--split-3</span> <span class="at">--skip-technical</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="ex">fasterq-dump</span> SRR_ID</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="downloading-reads-from-the-sra-using-fasterq-dump" class="level3">
<h3 class="anchored" data-anchor-id="downloading-reads-from-the-sra-using-fasterq-dump">Downloading reads from the SRA using <code>fasterq-dump</code></h3>
<p>In this example, we want to download FASTQ reads for a mate-pair library.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastq-dump</span> <span class="at">--threads</span> n <span class="at">--progress</span> SRR_ID</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>In this command…</p>
<ol type="1">
<li><strong><code>--threads</code></strong> specifies the number (<em><code>n</code></em>) processors/threads to be used.</li>
<li><strong><code>--progress</code></strong> is an optional argument that displays a progress bar when the reads are being downloaded.</li>
<li><strong><code>SRR_ID</code></strong> is the ID of the run from the SRA to be downloaded. This ID begins with “SRR” and is followed by around seven digits (e.g.&nbsp;<code>SRR1234567</code>).</li>
</ol>



</section>
</section>
</section>


<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-khoroshevskyi_geofetch:_2023" class="csl-entry" role="listitem">
Khoroshevskyi, Oleksandr, Nathan LeRoy, Vincent P Reuter, and Nathan C Sheffield. 2023. <span>“<span>GEOfetch</span>: A Command-Line Tool for Downloading Data and Standardized Metadata from <span>GEO</span> and <span>SRA</span>.”</span> Edited by Can Alkan. <em>Bioinformatics</em> 39 (3): btad069. <a href="https://doi.org/10.1093/bioinformatics/btad069">https://doi.org/10.1093/bioinformatics/btad069</a>.
</div>
<div id="ref-leroy_pephub:_2024" class="csl-entry" role="listitem">
LeRoy, Nathan J, Oleksandr Khoroshevskyi, Aaron O’Brien, Rafał Stępień, Alip Arslan, and Nathan C Sheffield. 2024. <span>“<span>PEPhub</span>: A Database, Web Interface, and <span>API</span> for Editing, Sharing, and Validating Biological Sample Metadata.”</span> <em>GigaScience</em> 13 (January): giae033. <a href="https://doi.org/10.1093/gigascience/giae033">https://doi.org/10.1093/gigascience/giae033</a>.
</div>
<div id="ref-sheffield_linking_2021" class="csl-entry" role="listitem">
Sheffield, Nathan C, Michał Stolarczyk, Vincent P Reuter, and André F Rendeiro. 2021. <span>“Linking Big Biomedical Datasets to Modular Analysis with <span>Portable</span> <span>Encapsulated</span> <span>Projects</span>.”</span> <em>GigaScience</em> 10 (12): giab077. <a href="https://doi.org/10.1093/gigascience/giab077">https://doi.org/10.1093/gigascience/giab077</a>.
</div>
</div></section><aside id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>FTP stands for File Transfer Protocol, and consists of a remote folder structure which you can access with URLs starting with <code>ftp://</code><a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</aside></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    const typesetMath = (el) => {
      if (window.MathJax) {
        // MathJax Typeset
        window.MathJax.typeset([el]);
      } else if (window.katex) {
        // KaTeX Render
        var mathElements = el.getElementsByClassName("math");
        var macros = [];
        for (var i = 0; i < mathElements.length; i++) {
          var texText = mathElements[i].firstChild;
          if (mathElements[i].tagName == "SPAN") {
            window.katex.render(texText.data, mathElements[i], {
              displayMode: mathElements[i].classList.contains('display'),
              throwOnError: false,
              macros: macros,
              fleqn: false
            });
          }
        }
      }
    }
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        typesetMath(container);
        return container.innerHTML
      } else {
        typesetMath(note);
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      typesetMath(note);
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>