<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.515">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2025-03-13">
<meta name="description" content="Data format and analysis workflow for Single Cell Proteomics in R">

<title>ABC - Accessible Bioinformatics Cafe - ABC.14: Single Cell Proteomics Downstream Analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-WG01Z2LJ1J"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-WG01Z2LJ1J', { 'anonymize_ip': true});
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
<!-- MailerLite Universal -->
<script>
    (function(w,d,e,u,f,l,n){w[f]=w[f]||function(){(w[f].q=w[f].q||[])
    .push(arguments);},l=d.createElement(e),l.async=1,l.src=u,
    n=d.getElementsByTagName(e)[0],n.parentNode.insertBefore(l,n);})
    (window,document,'script','https://assets.mailerlite.com/js/universal.js','ml');
    ml('account', '1004787');
</script>
<!-- End MailerLite Universal -->

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta property="og:title" content="ABC - Accessible Bioinformatics Cafe - ABC.14: Single Cell Proteomics Downstream Analysis">
<meta property="og:description" content="Data format and analysis workflow for Single Cell Proteomics in R">
<meta property="og:image" content="https://abc.au.dk
litesrv._domainkey.birc.au.dk/img/news/SCProteomics.png">
<meta property="og:site_name" content="ABC - Accessible Bioinformatics Cafe">
<meta property="og:locale" content="en_US">
<meta property="og:image:height" content="4783">
<meta property="og:image:width" content="4147">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../img/ABClogo.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ABC - Accessible Bioinformatics Cafe</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../news.html"> 
<span class="menu-text">Calendar</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Documentation.html"> 
<span class="menu-text">Documentation</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-resources" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Resources</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-resources">    
        <li>
    <a class="dropdown-item" href="../binfResources/bash.html">
 <span class="dropdown-text">Bash tips</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../gethelp.html"> 
<span class="menu-text">People&amp;Help</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/AU-ABC"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#slides" id="toc-slides" class="nav-link active" data-scroll-target="#slides">Slides</a></li>
  <li><a href="#tutorial" id="toc-tutorial" class="nav-link" data-scroll-target="#tutorial">Tutorial</a>
  <ul class="collapse">
  <li><a href="#how-to-run-the-tutorial" id="toc-how-to-run-the-tutorial" class="nav-link" data-scroll-target="#how-to-run-the-tutorial">How to run the tutorial</a>
  <ul class="collapse">
  <li><a href="#from-your-computer" id="toc-from-your-computer" class="nav-link" data-scroll-target="#from-your-computer">From your computer</a></li>
  <li><a href="#with-pixi-on-the-cluster" id="toc-with-pixi-on-the-cluster" class="nav-link" data-scroll-target="#with-pixi-on-the-cluster">With pixi on the cluster</a></li>
  <li><a href="#with-conda-on-the-cluster" id="toc-with-conda-on-the-cluster" class="nav-link" data-scroll-target="#with-conda-on-the-cluster">With conda on the cluster</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#the-qfeatures-class" id="toc-the-qfeatures-class" class="nav-link" data-scroll-target="#the-qfeatures-class">The <code>QFeatures</code> class</a></li>
  <li><a href="#accessing-the-data" id="toc-accessing-the-data" class="nav-link" data-scroll-target="#accessing-the-data">Accessing the data</a>
  <ul class="collapse">
  <li><a href="#quantitative-data" id="toc-quantitative-data" class="nav-link" data-scroll-target="#quantitative-data">Quantitative data</a></li>
  <li><a href="#feature-metadata" id="toc-feature-metadata" class="nav-link" data-scroll-target="#feature-metadata">Feature metadata</a></li>
  <li><a href="#sample-metadata" id="toc-sample-metadata" class="nav-link" data-scroll-target="#sample-metadata">Sample metadata</a></li>
  </ul></li>
  <li><a href="#subsetting-the-data" id="toc-subsetting-the-data" class="nav-link" data-scroll-target="#subsetting-the-data">Subsetting the data</a>
  <ul class="collapse">
  <li><a href="#subset-assays" id="toc-subset-assays" class="nav-link" data-scroll-target="#subset-assays">Subset assays</a></li>
  <li><a href="#subset-samples" id="toc-subset-samples" class="nav-link" data-scroll-target="#subset-samples">Subset samples</a></li>
  <li><a href="#subset-features" id="toc-subset-features" class="nav-link" data-scroll-target="#subset-features">Subset features</a></li>
  </ul></li>
  <li><a href="#common-processing-steps" id="toc-common-processing-steps" class="nav-link" data-scroll-target="#common-processing-steps">Common processing steps</a>
  <ul class="collapse">
  <li><a href="#missing-data-assignment" id="toc-missing-data-assignment" class="nav-link" data-scroll-target="#missing-data-assignment">Missing data assignment</a></li>
  <li><a href="#feature-aggregation" id="toc-feature-aggregation" class="nav-link" data-scroll-target="#feature-aggregation">Feature aggregation</a></li>
  <li><a href="#normalization" id="toc-normalization" class="nav-link" data-scroll-target="#normalization">Normalization</a></li>
  <li><a href="#log-transformation" id="toc-log-transformation" class="nav-link" data-scroll-target="#log-transformation">Log transformation</a></li>
  <li><a href="#imputation" id="toc-imputation" class="nav-link" data-scroll-target="#imputation">Imputation</a></li>
  </ul></li>
  <li><a href="#data-visualization" id="toc-data-visualization" class="nav-link" data-scroll-target="#data-visualization">Data visualization</a></li>
  <li><a href="#single-cell-proteomics-data-processing-and-analysis" id="toc-single-cell-proteomics-data-processing-and-analysis" class="nav-link" data-scroll-target="#single-cell-proteomics-data-processing-and-analysis">Single Cell Proteomics data processing and analysis</a>
  <ul class="collapse">
  <li><a href="#the-scp-package" id="toc-the-scp-package" class="nav-link" data-scroll-target="#the-scp-package">The <code>scp</code> package</a></li>
  <li><a href="#before-you-start" id="toc-before-you-start" class="nav-link" data-scroll-target="#before-you-start">Before you start</a></li>
  <li><a href="#read-in-scp-data" id="toc-read-in-scp-data" class="nav-link" data-scroll-target="#read-in-scp-data">Read in SCP data</a></li>
  <li><a href="#quality-control" id="toc-quality-control" class="nav-link" data-scroll-target="#quality-control">Quality control</a>
  <ul class="collapse">
  <li><a href="#clean-missing-data" id="toc-clean-missing-data" class="nav-link" data-scroll-target="#clean-missing-data">Clean missing data</a></li>
  <li><a href="#filter-psms" id="toc-filter-psms" class="nav-link" data-scroll-target="#filter-psms">Filter PSMs</a></li>
  </ul></li>
  <li><a href="#process-the-psm-data" id="toc-process-the-psm-data" class="nav-link" data-scroll-target="#process-the-psm-data">Process the PSM data</a>
  <ul class="collapse">
  <li><a href="#relative-reporter-ion-intensity" id="toc-relative-reporter-ion-intensity" class="nav-link" data-scroll-target="#relative-reporter-ion-intensity">Relative reporter ion intensity</a></li>
  </ul></li>
  <li><a href="#aggregate-psm-data-to-peptide-data" id="toc-aggregate-psm-data-to-peptide-data" class="nav-link" data-scroll-target="#aggregate-psm-data-to-peptide-data">Aggregate PSM data to peptide data</a></li>
  <li><a href="#join-the-scope2-sets-in-one-assay" id="toc-join-the-scope2-sets-in-one-assay" class="nav-link" data-scroll-target="#join-the-scope2-sets-in-one-assay">Join the SCoPE2 sets in one assay</a></li>
  <li><a href="#filter-single-cells" id="toc-filter-single-cells" class="nav-link" data-scroll-target="#filter-single-cells">Filter single-cells</a>
  <ul class="collapse">
  <li><a href="#filter-samples-of-interest" id="toc-filter-samples-of-interest" class="nav-link" data-scroll-target="#filter-samples-of-interest">Filter samples of interest</a></li>
  <li><a href="#filter-based-on-the-median-relative-intensity" id="toc-filter-based-on-the-median-relative-intensity" class="nav-link" data-scroll-target="#filter-based-on-the-median-relative-intensity">Filter based on the median relative intensity</a></li>
  <li><a href="#filter-based-on-the-median-cv" id="toc-filter-based-on-the-median-cv" class="nav-link" data-scroll-target="#filter-based-on-the-median-cv">Filter based on the median CV</a></li>
  </ul></li>
  <li><a href="#process-the-peptide-data" id="toc-process-the-peptide-data" class="nav-link" data-scroll-target="#process-the-peptide-data">Process the peptide data</a>
  <ul class="collapse">
  <li><a href="#normalization-1" id="toc-normalization-1" class="nav-link" data-scroll-target="#normalization-1">Normalization</a></li>
  <li><a href="#remove-peptides-with-high-missing-rate" id="toc-remove-peptides-with-high-missing-rate" class="nav-link" data-scroll-target="#remove-peptides-with-high-missing-rate">Remove peptides with high missing rate</a></li>
  <li><a href="#log-transformation-1" id="toc-log-transformation-1" class="nav-link" data-scroll-target="#log-transformation-1">Log-transformation</a></li>
  </ul></li>
  <li><a href="#aggregate-peptide-data-to-protein-data" id="toc-aggregate-peptide-data-to-protein-data" class="nav-link" data-scroll-target="#aggregate-peptide-data-to-protein-data">Aggregate peptide data to protein data</a></li>
  <li><a href="#process-the-protein-data" id="toc-process-the-protein-data" class="nav-link" data-scroll-target="#process-the-protein-data">Process the protein data</a>
  <ul class="collapse">
  <li><a href="#normalization-2" id="toc-normalization-2" class="nav-link" data-scroll-target="#normalization-2">Normalization</a></li>
  <li><a href="#imputation-1" id="toc-imputation-1" class="nav-link" data-scroll-target="#imputation-1">Imputation</a></li>
  <li><a href="#batch-correction" id="toc-batch-correction" class="nav-link" data-scroll-target="#batch-correction">Batch correction</a></li>
  </ul></li>
  <li><a href="#dimension-reduction" id="toc-dimension-reduction" class="nav-link" data-scroll-target="#dimension-reduction">Dimension reduction</a>
  <ul class="collapse">
  <li><a href="#pca" id="toc-pca" class="nav-link" data-scroll-target="#pca">PCA</a></li>
  <li><a href="#umap" id="toc-umap" class="nav-link" data-scroll-target="#umap">UMAP</a></li>
  </ul></li>
  <li><a href="#monitoring-data-processing" id="toc-monitoring-data-processing" class="nav-link" data-scroll-target="#monitoring-data-processing">Monitoring data processing</a></li>
  </ul></li>
  <li><a href="#session-information" id="toc-session-information" class="nav-link" data-scroll-target="#session-information">Session information</a></li>
  <li><a href="#license" id="toc-license" class="nav-link" data-scroll-target="#license">License</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">ABC.14: Single Cell Proteomics Downstream Analysis</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Single Cell Proteomics</div>
    <div class="quarto-category">SCP</div>
    <div class="quarto-category">Mass Spectrometry</div>
    <div class="quarto-category">QFeatures</div>
    <div class="quarto-category">Data Formats</div>
  </div>
  </div>

<div>
  <div class="description">
    Data format and analysis workflow for Single Cell Proteomics in R
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Authors</div>
    <div class="quarto-title-meta-contents">
             <p>Samuele Soraggi </p>
             <p>Laurent Gatto </p>
             <p>Cristophe Vanderaa </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">March 13, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="slides" class="level1">
<h1>Slides</h1>
<p>Today’s slides with topic presentation for Single Cell Proteomics downstream analysis with SC proteomics data obtained through Mass Spectrometry.&nbsp;</p>
<p>&nbsp;</p>
<p align="center">
</p><p><a href="https://abc.au.dk/documentation/slides/20250313-ABC14.zip" style="background-color: #4266A1; color: #FFFFFF; padding: 30px 20px; text-decoration: none; border-radius: 5px;"> Download Slides </a></p>
<p></p>
<p>&nbsp;</p>
</section>
<section id="tutorial" class="level1">
<h1>Tutorial</h1>
<p>This tutorial is extracted from multiple vignettes of the <code>R</code> package <code>scp</code> and <code>QFeatures</code>. The <code>scp</code> package is a collection of tools for the analysis of single-cell proteomics data, and the <code>QFeatures</code> package is a collection of tools for the analysis of quantitative features in mass spectrometry-based proteomics data.</p>
<p>The original authors of the code are Laurent Gatto and Cristophe Vanderaa, which we edited to suit the tutorial format. The code is licensed as per the authors’ license (see bottom of this page).</p>
<div class="callout callout-style-default callout-note callout-titled" title="Learning Objectives">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Understand the <code>QFeatures</code> class and its structure</li>
<li>Learn how to load and manipulate <code>QFeatures</code> objects</li>
<li>Learn how to visualize <code>QFeatures</code> objects</li>
<li>Preprocess features and cells single-cell proteomics data</li>
<li>Perform operations on single-cell proteomics data</li>
</ul>
</div>
</div>
<section id="how-to-run-the-tutorial" class="level2">
<h2 class="anchored" data-anchor-id="how-to-run-the-tutorial">How to run the tutorial</h2>
<section id="from-your-computer" class="level3">
<h3 class="anchored" data-anchor-id="from-your-computer">From your computer</h3>
<p>You can run this tutorial on your laptop by installing the required packages <code>scp</code>, <code>dplyr</code> and <code>ggplot2</code>.</p>
</section>
<section id="with-pixi-on-the-cluster" class="level3">
<h3 class="anchored" data-anchor-id="with-pixi-on-the-cluster">With pixi on the cluster</h3>
<p>You can also run the tutorial from <code>Rstudio</code> on GenomeDK, but only if you use the command line on the virtual desktop at <code>desktop.genome.au.dk</code>. First, log into the virtual desktop by filling in name and password for GenomeDK at the webpage <code>desktop.genome.au.dk</code>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="sharedImages/GDKDesktopLogin.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="400"></p>
</figure>
</div>
<p>Then, use the file browser (red circle) to navigate into a project folder where you want to work. Create an empty folder for the tutorial called ABC14 and open it. You can create a folder with the right-click menu (green circle).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="sharedImages/GDKDesktopRightClick.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="800"></p>
</figure>
</div>
<p>Once you are in the new folder, use the right-click menu to create an empty document. Rename it <code>pixi.toml</code>, open it, and paste the code block below. Save the file and close it.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode toml code-with-copy"><code class="sourceCode toml"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="kw">[project]</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="dt">authors</span> <span class="op">=</span> <span class="op">[</span><span class="st">"SamueleSoraggi &lt;samuele@birc.au.dk&gt;"</span><span class="op">]</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="dt">channels</span> <span class="op">=</span> <span class="op">[</span><span class="st">"conda-forge"</span><span class="op">,</span> <span class="st">"bioconda"</span><span class="op">]</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="dt">name</span> <span class="op">=</span> <span class="st">"ABC14-SCProteomics"</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="dt">platforms</span> <span class="op">=</span> <span class="op">[</span><span class="st">"linux-64"</span><span class="op">]</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="dt">version</span> <span class="op">=</span> <span class="st">"0.1.0"</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="kw">[tasks]</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="kw">[dependencies]</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="dt">jupyterlab</span> <span class="op">=</span> <span class="st">"&gt;=4.3.5,&lt;5"</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="dt">rstudio-desktop</span> <span class="op">=</span> <span class="st">"&gt;=2024.4.2,&lt;2025"</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="dt">bioconductor-scp</span> <span class="op">=</span> <span class="st">"&gt;=1.16.0,&lt;2"</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="dt">r-ggplot2</span> <span class="op">=</span> <span class="st">"&gt;=3.5.1,&lt;4"</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="dt">r-dplyr</span> <span class="op">=</span> <span class="st">"&gt;=1.1.4,&lt;2"</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="dt">r-patchwork</span> <span class="op">=</span> <span class="st">"&gt;=1.3.0,&lt;2"</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="dt">bioconductor-singlecellexperiment</span> <span class="op">=</span> <span class="st">"&gt;=1.28.0,&lt;2"</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="dt">r-irkernel</span> <span class="op">=</span> <span class="st">"&gt;=1.3.2,&lt;2"</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="dt">bioconductor-genomeinfodb</span> <span class="op">=</span> <span class="st">"&gt;=1.42.0,&lt;2"</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="dt">bioconductor-genomeinfodbdata</span> <span class="op">=</span> <span class="st">"&gt;=1.2.13,&lt;2"</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="dt">bioconductor-impute</span> <span class="op">=</span> <span class="st">"&gt;=1.80.0,&lt;2"</span></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="dt">pip</span> <span class="op">=</span> <span class="st">"&gt;=25.0.1,&lt;26"</span></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="dt">bioconductor-sva</span> <span class="op">=</span> <span class="st">"&gt;=3.54.0,&lt;4"</span></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a><span class="dt">bioconductor-scater</span> <span class="op">=</span> <span class="st">"&gt;=1.34.0,&lt;2"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Use the right-click menu again and choose <code>Open terminal here</code>. A terminal already logged into GenomeDK will open.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="sharedImages/GDKDesktopCLI.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="300"></p>
</figure>
</div>
<p>If you do not have <code>pixi installed</code>, follow the few steps in the hidden box below, otherwise skip to the next steps.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Install pixi">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Install pixi
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Run the installation script to install the software. There are some informative messages during the installation.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-fsSL</span> https://pixi.sh/install.sh <span class="kw">|</span> <span class="fu">bash</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>When you are done, you need a little bit of configuration. You can run the following command to configure <code>pixi</code> (run them only once, they are not needed again):</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> PATH=<span class="st">"</span><span class="va">$PATH</span><span class="st">:</span><span class="va">$HOME</span><span class="st">/.pixi/bin"</span> <span class="op">&gt;&gt;</span> ~/.bashrc</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">'eval "$(pixi completion --shell bash)"'</span>  <span class="op">&gt;&gt;</span> ~/.bashrc</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/.bashrc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now you are ready to move on in the tutorial!</p>
</div>
</div>
</div>
<p>In the terminal, start a new interactive job with a command like the one below</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--mem</span><span class="op">=</span>16g <span class="at">--cores</span><span class="op">=</span>2 <span class="at">--time</span><span class="op">=</span>02:00:00 <span class="at">--account</span><span class="op">=</span>PROJECT <span class="at">--pty</span> /bin/bash</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Choose the amount of memory, cores and hours of usage you need. You also need a project name to write instead of <code>--account=PROJECT</code>. If you have no projects available, simply delete that option.</p>
<p>Once the waiting time for the interactive job is finished, use the command <code>pixi run rstudio</code>. The packages will be installed and <code>Rstudio</code> will open!</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>One you have opened <code>Rstudio</code>, you also need to manually install some extra bioconductor packages using the <code>R</code> console:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"GenomeInfoDbData"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="with-conda-on-the-cluster" class="level3">
<h3 class="anchored" data-anchor-id="with-conda-on-the-cluster">With conda on the cluster</h3>
<p>Repeat all the steps described above until the creation of the <code>pixi.toml</code> file. Instead, create a <code>environment.yml</code> file with the following content:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">channels</span><span class="kw">:</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> conda-forge</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> bioconda</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dependencies</span><span class="kw">:</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> jupyterlab &gt;=4.3.5,&lt;5</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> rstudio-desktop &gt;=2024.4.2,&lt;2025</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> bioconductor-scp &gt;=1.16.0,&lt;2</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> r-ggplot2 &gt;=3.5.1,&lt;4</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> r-dplyr &gt;=1.1.4,&lt;2</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> r-patchwork &gt;=1.3.0,&lt;2</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> bioconductor-singlecellexperiment &gt;=1.28.0,&lt;2</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> r-irkernel &gt;=1.3.2,&lt;2</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> bioconductor-genomeinfodb &gt;=1.42.0,&lt;2</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> bioconductor-genomeinfodbdata &gt;=1.2.13,&lt;2</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> bioconductor-impute &gt;=1.80.0,&lt;2</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> pip &gt;=25.0.1,&lt;26</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> bioconductor-sva &gt;=3.54.0,&lt;4</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="kw">-</span><span class="at"> bioconductor-scater &gt;=1.34.0,&lt;2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Use the right-click menu again and choose <code>Open terminal here</code>. A terminal already logged into GenomeDK will open.</p>
<p>If you do not have <code>conda</code> installed, follow the few steps in the hidden box below, otherwise skip to the next steps.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Install conda">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Install conda
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Run the installation script to install the software. There are some informative messages during the installation. You might need to say <code>yes</code> a few times</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-x86_64.sh <span class="at">-O</span> miniforge.sh</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">chmod</span> +x miniforge.sh</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">bash</span> miniforge.sh <span class="at">-b</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="ex">~/miniforge3/bin/conda</span> init bash</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/.bashrc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>When you are done, you need a little bit of configuration. You can run the following command to configure <code>conda</code> (run them only once, they are not needed again):</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> config <span class="at">--append</span> channels conda-forge</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> config <span class="at">--append</span> channels bioconda</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> config <span class="at">--set</span> channel_priority strict</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> config <span class="at">--set</span> auto_activate_base false</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/.bashrc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Finally, install a package to accellerate <code>conda</code>. This is optional but recommended:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-n</span> base <span class="at">--yes</span> conda-libmamba-solver</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> config <span class="at">--set</span> solver libmamba</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now you are ready to move on in the tutorial!</p>
</div>
</div>
</div>
<p>In the terminal, create a new conda named <code>ABC14</code> environment with the command below. Be patient because it takes some minutes:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> env create <span class="at">-f</span> environment.yml <span class="at">-n</span> ABC14</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Then start a new interactive job with a command like the one below</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--mem</span><span class="op">=</span>16g <span class="at">--cores</span><span class="op">=</span>2 <span class="at">--time</span><span class="op">=</span>02:00:00 <span class="at">--account</span><span class="op">=</span>PROJECT <span class="at">--pty</span> /bin/bash</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Choose the amount of memory, cores and hours of usage you need. You also need a project name to write instead of <code>--account=PROJECT</code>. If you have no projects available, simply delete that option.</p>
<p>Once the waiting time for the interactive job is finished, use the command <code>conda activate ABC14</code> to make all installed packages available. Then use the command <code>rstudio</code> to open <code>Rstudio</code>!</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>One you have opened <code>Rstudio</code>, you also need to manually install some extra bioconductor packages using the <code>R</code> console:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"GenomeInfoDbData"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
</section>
</section>
<section id="the-qfeatures-class" class="level1">
<h1>The <code>QFeatures</code> class</h1>
<p>The <code>QFeatures</code> class (Gatto and Vanderaa (2023)) is based on the <code>MultiAssayExperiment</code> class that holds a collection of <code>SummarizedExperiment</code> (or other classes that inherits from it) objects termed <strong>assays</strong>. The assays in a <code>QFeatures</code> object have a hierarchical relation: proteins are composed of peptides, themselves produced by spectra, as depicted in figure below.</p>
<p><img src="https://rformassspectrometry.github.io/QFeatures/articles/QFeatures_files/figure-html/featuresplot-1.png" class="img-fluid"></p>
<p>A more technical representation is shown below, highlighting that each assay is a <code>SummarizedExperiment</code> (containing the quantitative data, row and column annotations for each individual assay), as well as a global sample annotation table, that annotates cells across all assays.</p>
<p><img src="https://lgatto.github.io/QFeaturesScpWorkshop2021/articles/figs/SCP_framework.png" class="img-fluid"></p>
<p>Those links are stored as part as the <code>QFeatures</code> object and connect the assays together. We load an example dataset from the <code>scp</code> package that is formatted as an <code>QFeatures</code> object and plot those connection.</p>
<div id="523f95df-86ad-4c2c-9a05-791b64b49cba" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scp)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"scp1"</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(scp1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: QFeatures

Loading required package: MultiAssayExperiment

Loading required package: SummarizedExperiment

Loading required package: MatrixGenerics

Loading required package: matrixStats


Attaching package: ‘MatrixGenerics’


The following objects are masked from ‘package:matrixStats’:

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars


Loading required package: GenomicRanges

Loading required package: stats4

Loading required package: BiocGenerics


Attaching package: ‘BiocGenerics’


The following objects are masked from ‘package:stats’:

    IQR, mad, sd, var, xtabs


The following objects are masked from ‘package:base’:

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,
    table, tapply, union, unique, unsplit, which.max, which.min


Loading required package: S4Vectors


Attaching package: ‘S4Vectors’


The following object is masked from ‘package:utils’:

    findMatches


The following objects are masked from ‘package:base’:

    expand.grid, I, unname


Loading required package: IRanges

Loading required package: GenomeInfoDb

Loading required package: Biobase

Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.



Attaching package: ‘Biobase’


The following object is masked from ‘package:MatrixGenerics’:

    rowMedians


The following objects are masked from ‘package:matrixStats’:

    anyMissing, rowMedians



Attaching package: ‘QFeatures’


The following object is masked from ‘package:MultiAssayExperiment’:

    longFormat


The following object is masked from ‘package:base’:

    sweep

</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-03-13-ABC14_files/figure-html/cell-2-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="accessing-the-data" class="level1">
<h1>Accessing the data</h1>
<p>The <code>QFeatures</code> class contains all the available and metadata. We here show how to retrieve those different pieces of information.</p>
<section id="quantitative-data" class="level2">
<h2 class="anchored" data-anchor-id="quantitative-data">Quantitative data</h2>
<p>The quantitative data, stored as matrix-like objects, can be accessed using the <code>assay</code> function. For example, we here extract the quantitative data for the first MS batch (and show a subset of it):</p>
<div id="c12b25ab-762d-4319-89aa-f427f66faebd" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="2">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(scp1, <span class="st">"190321S_LCA10_X_FP97AG"</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A matrix: 5 × 11 of type dbl</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">190321S_LCA10_X_FP97AG_RI1</th>
<th data-quarto-table-cell-role="th" scope="col">190321S_LCA10_X_FP97AG_RI2</th>
<th data-quarto-table-cell-role="th" scope="col">190321S_LCA10_X_FP97AG_RI3</th>
<th data-quarto-table-cell-role="th" scope="col">190321S_LCA10_X_FP97AG_RI4</th>
<th data-quarto-table-cell-role="th" scope="col">190321S_LCA10_X_FP97AG_RI5</th>
<th data-quarto-table-cell-role="th" scope="col">190321S_LCA10_X_FP97AG_RI6</th>
<th data-quarto-table-cell-role="th" scope="col">190321S_LCA10_X_FP97AG_RI7</th>
<th data-quarto-table-cell-role="th" scope="col">190321S_LCA10_X_FP97AG_RI8</th>
<th data-quarto-table-cell-role="th" scope="col">190321S_LCA10_X_FP97AG_RI9</th>
<th data-quarto-table-cell-role="th" scope="col">190321S_LCA10_X_FP97AG_RI10</th>
<th data-quarto-table-cell-role="th" scope="col">190321S_LCA10_X_FP97AG_RI11</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">PSM3773</td>
<td>57895</td>
<td>603.73</td>
<td>2787.9</td>
<td>757.17</td>
<td>862.08</td>
<td>1118.80</td>
<td>640.10</td>
<td>1446.10</td>
<td>968.49</td>
<td>648.56</td>
<td>742.53</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">PSM9078</td>
<td>64889</td>
<td>1481.30</td>
<td>4891.6</td>
<td>597.53</td>
<td>1140.30</td>
<td>1300.10</td>
<td>1092.50</td>
<td>1309.40</td>
<td>1538.40</td>
<td>1014.50</td>
<td>1062.80</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">PSM9858</td>
<td>58993</td>
<td>489.85</td>
<td>2899.4</td>
<td>882.37</td>
<td>296.60</td>
<td>977.15</td>
<td>498.60</td>
<td>1437.90</td>
<td>857.40</td>
<td>888.01</td>
<td>768.61</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">PSM11744</td>
<td>75711</td>
<td>539.02</td>
<td>7292.7</td>
<td>357.90</td>
<td>1091.30</td>
<td>736.87</td>
<td>712.74</td>
<td>590.75</td>
<td>15623.00</td>
<td>298.60</td>
<td>481.38</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">PSM21752</td>
<td>0</td>
<td>0.00</td>
<td>0.0</td>
<td>0.00</td>
<td>0.00</td>
<td>0.00</td>
<td>0.00</td>
<td>0.00</td>
<td>0.00</td>
<td>0.00</td>
<td>0.00</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Note that you can retrieve the list of available assays in a <code>QFeatures</code> object using the <code>names()</code> function.</p>
<div id="0df0eeda-becf-4bfb-a4ed-e5f15b27bc38" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="3">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(scp1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>'190321S_LCA10_X_FP97AG'</li><li>'190222S_LCA9_X_FP94BM'</li><li>'190914S_LCB3_X_16plex_Set_21'</li><li>'peptides'</li><li>'proteins'</li></ol>
</div>
</div>
</section>
<section id="feature-metadata" class="level2">
<h2 class="anchored" data-anchor-id="feature-metadata">Feature metadata</h2>
<p>For each individual assay, there is feature metadata available. We extract the list of metadata tables by using <code>rowData()</code> on the <code>QFeatures</code> object.</p>
<div id="8ca85c57-082c-473e-bfa1-821fc49e2dee" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="4">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(scp1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>DataFrameList of length 5
names(5): 190321S_LCA10_X_FP97AG 190222S_LCA9_X_FP94BM 190914S_LCB3_X_16plex_Set_21 peptides proteins</code></pre>
</div>
</div>
<div id="14d1c4c7-644c-413b-9bc7-a65e0642380d" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="5">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(scp1)[[<span class="st">"proteins"</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>DataFrame with 292 rows and 9 columns
                                           protein Match.time.difference
                                       &lt;character&gt;             &lt;logical&gt;
A1A519                                      A1A519                    NA
A5D8V6                                      A5D8V6                    NA
A5PLK6                                      A5PLK6                    NA
A5PLL1                                      A5PLL1                    NA
A6NC97                                      A6NC97                    NA
...                                            ...                   ...
REV__CON__ENSEMBL:ENSBTAP00000038253 REV__CON__...                    NA
REV__CON__P06868                     REV__CON__...                    NA
REV__CON__Q05443                     REV__CON__...                    NA
REV__CON__Q32PI4                     REV__CON__...                    NA
REV__CON__Q3MHN5                     REV__CON__...                    NA
                                     Match.m.z.difference Match.q.value
                                                &lt;logical&gt;     &lt;logical&gt;
A1A519                                                 NA            NA
A5D8V6                                                 NA            NA
A5PLK6                                                 NA            NA
A5PLL1                                                 NA            NA
A6NC97                                                 NA            NA
...                                                   ...           ...
REV__CON__ENSEMBL:ENSBTAP00000038253                   NA            NA
REV__CON__P06868                                       NA            NA
REV__CON__Q05443                                       NA            NA
REV__CON__Q32PI4                                       NA            NA
REV__CON__Q3MHN5                                       NA            NA
                                     Match.score Reporter.PIF Reporter.fraction
                                       &lt;logical&gt;    &lt;logical&gt;         &lt;logical&gt;
A1A519                                        NA           NA                NA
A5D8V6                                        NA           NA                NA
A5PLK6                                        NA           NA                NA
A5PLL1                                        NA           NA                NA
A6NC97                                        NA           NA                NA
...                                          ...          ...               ...
REV__CON__ENSEMBL:ENSBTAP00000038253          NA           NA                NA
REV__CON__P06868                              NA           NA                NA
REV__CON__Q05443                              NA           NA                NA
REV__CON__Q32PI4                              NA           NA                NA
REV__CON__Q3MHN5                              NA           NA                NA
                                     Potential.contaminant        .n
                                               &lt;character&gt; &lt;integer&gt;
A1A519                                                             1
A5D8V6                                                             1
A5PLK6                                                             1
A5PLL1                                                             1
A6NC97                                                             1
...                                                    ...       ...
REV__CON__ENSEMBL:ENSBTAP00000038253                     +         1
REV__CON__P06868                                         +         1
REV__CON__Q05443                                         +         1
REV__CON__Q32PI4                                         +         1
REV__CON__Q3MHN5                                         +         1</code></pre>
</div>
</div>
<p>You can also retrieve the names of each <code>rowData</code> column for all assays with <code>rowDataNames</code>.</p>
<div id="9ae41fd4-107e-4378-ab9e-c5846b47670d" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="6">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowDataNames</span>(scp1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>CharacterList of length 5
[["190321S_LCA10_X_FP97AG"]] uid Sequence ... peptide Leading.razor.protein
[["190222S_LCA9_X_FP94BM"]] uid Sequence ... peptide Leading.razor.protein
[["190914S_LCB3_X_16plex_Set_21"]] uid Sequence ... Leading.razor.protein
[["peptides"]] Sequence Length Modifications ... .n Leading.razor.protein
[["proteins"]] protein Match.time.difference ... Potential.contaminant .n</code></pre>
</div>
</div>
<p>You can also get the <code>rowData</code> from different assays in a single table using the <code>rbindRowData</code> function. It will keep the common <code>rowData</code> variables to all selected assays (provided through <code>i</code>).</p>
<div id="56792c9b-c60c-4712-a213-bd4812f760a7" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="7">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rbindRowData</span>(scp1, <span class="at">i =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>DataFrame with 1388 rows and 10 columns
             assay       rowname       protein Match.time.difference
       &lt;character&gt;   &lt;character&gt;   &lt;character&gt;             &lt;logical&gt;
1    190321S_LC...       PSM3773        P61981                    NA
2    190321S_LC...       PSM9078        Q8WVN8                    NA
3    190321S_LC...       PSM9858        P55084                    NA
4    190321S_LC...      PSM11744        P19099                    NA
5    190321S_LC...      PSM21752        P52952                    NA
...            ...           ...           ...                   ...
1384      proteins REV__CON__... REV__CON__...                    NA
1385      proteins REV__CON__... REV__CON__...                    NA
1386      proteins REV__CON__... REV__CON__...                    NA
1387      proteins REV__CON__... REV__CON__...                    NA
1388      proteins REV__CON__... REV__CON__...                    NA
     Match.m.z.difference Match.q.value Match.score Reporter.PIF
                &lt;logical&gt;     &lt;logical&gt;   &lt;logical&gt;    &lt;logical&gt;
1                      NA            NA          NA           NA
2                      NA            NA          NA           NA
3                      NA            NA          NA           NA
4                      NA            NA          NA           NA
5                      NA            NA          NA           NA
...                   ...           ...         ...          ...
1384                   NA            NA          NA           NA
1385                   NA            NA          NA           NA
1386                   NA            NA          NA           NA
1387                   NA            NA          NA           NA
1388                   NA            NA          NA           NA
     Reporter.fraction Potential.contaminant
             &lt;logical&gt;           &lt;character&gt;
1                   NA                      
2                   NA                      
3                   NA                      
4                   NA                      
5                   NA                      
...                ...                   ...
1384                NA                     +
1385                NA                     +
1386                NA                     +
1387                NA                     +
1388                NA                     +</code></pre>
</div>
</div>
</section>
<section id="sample-metadata" class="level2">
<h2 class="anchored" data-anchor-id="sample-metadata">Sample metadata</h2>
<p>The sample metadata is retrieved using <code>colData</code> on the <code>QFeatures</code> object.</p>
<div id="26f209ba-ed4f-40a5-851a-732215218b8d" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="8">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(scp1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>DataFrame with 38 rows and 7 columns
                                            Set     Channel SampleAnnotation
                                    &lt;character&gt; &lt;character&gt;      &lt;character&gt;
190222S_LCA9_X_FP94BM_RI1         190222S_LC...         RI1    carrier_mi...
190222S_LCA9_X_FP94BM_RI2         190222S_LC...         RI2             norm
190222S_LCA9_X_FP94BM_RI3         190222S_LC...         RI3           unused
190222S_LCA9_X_FP94BM_RI4         190222S_LC...         RI4             sc_u
190222S_LCA9_X_FP94BM_RI5         190222S_LC...         RI5             sc_0
...                                         ...         ...              ...
190914S_LCB3_X_16plex_Set_21_RI12 190914S_LC...        RI12            sc_m0
190914S_LCB3_X_16plex_Set_21_RI13 190914S_LC...        RI13            sc_m0
190914S_LCB3_X_16plex_Set_21_RI14 190914S_LC...        RI14            sc_m0
190914S_LCB3_X_16plex_Set_21_RI15 190914S_LC...        RI15            sc_m0
190914S_LCB3_X_16plex_Set_21_RI16 190914S_LC...        RI16            sc_m0
                                   SampleType     lcbatch     sortday
                                  &lt;character&gt; &lt;character&gt; &lt;character&gt;
190222S_LCA9_X_FP94BM_RI1             Carrier        LCA9          s8
190222S_LCA9_X_FP94BM_RI2           Reference        LCA9          s8
190222S_LCA9_X_FP94BM_RI3              Unused        LCA9          s8
190222S_LCA9_X_FP94BM_RI4            Monocyte        LCA9          s8
190222S_LCA9_X_FP94BM_RI5               Blank        LCA9          s8
...                                       ...         ...         ...
190914S_LCB3_X_16plex_Set_21_RI12  Macrophage        LCB3          s9
190914S_LCB3_X_16plex_Set_21_RI13  Macrophage        LCB3          s9
190914S_LCB3_X_16plex_Set_21_RI14  Macrophage        LCB3          s9
190914S_LCB3_X_16plex_Set_21_RI15  Macrophage        LCB3          s9
190914S_LCB3_X_16plex_Set_21_RI16  Macrophage        LCB3          s9
                                       digest
                                  &lt;character&gt;
190222S_LCA9_X_FP94BM_RI1                   N
190222S_LCA9_X_FP94BM_RI2                   N
190222S_LCA9_X_FP94BM_RI3                   N
190222S_LCA9_X_FP94BM_RI4                   N
190222S_LCA9_X_FP94BM_RI5                   N
...                                       ...
190914S_LCB3_X_16plex_Set_21_RI12           R
190914S_LCB3_X_16plex_Set_21_RI13           R
190914S_LCB3_X_16plex_Set_21_RI14           R
190914S_LCB3_X_16plex_Set_21_RI15           R
190914S_LCB3_X_16plex_Set_21_RI16           R</code></pre>
</div>
</div>
<p>Note that you can easily access a <code>colData</code> column using the <code>$</code> operator. See here how we extract the sample types from the <code>colData</code>.</p>
<div id="7d919dc4-a72c-418b-b4fb-627cb547acc6" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="9">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>scp1<span class="sc">$</span>SampleType</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>'Carrier'</li><li>'Reference'</li><li>'Unused'</li><li>'Monocyte'</li><li>'Blank'</li><li>'Monocyte'</li><li>'Macrophage'</li><li>'Macrophage'</li><li>'Macrophage'</li><li>'Macrophage'</li><li>'Macrophage'</li><li>'Carrier'</li><li>'Reference'</li><li>'Unused'</li><li>'Macrophage'</li><li>'Monocyte'</li><li>'Macrophage'</li><li>'Macrophage'</li><li>'Macrophage'</li><li>'Macrophage'</li><li>'Macrophage'</li><li>'Macrophage'</li><li>'Carrier'</li><li>'Reference'</li><li>'Unused'</li><li>'Unused'</li><li>'Macrophage'</li><li>'Macrophage'</li><li>'Blank'</li><li>'Monocyte'</li><li>'Macrophage'</li><li>'Monocyte'</li><li>'Blank'</li><li>'Macrophage'</li><li>'Macrophage'</li><li>'Macrophage'</li><li>'Macrophage'</li><li>'Macrophage'</li></ol>
</div>
</div>
</section>
</section>
<section id="subsetting-the-data" class="level1">
<h1>Subsetting the data</h1>
<p>There are three dimensions we want to subset for:</p>
<ul>
<li>Assays</li>
<li>Samples</li>
<li>Features</li>
</ul>
<p>Therefore, <code>QFeatures</code> support a three-index subsetting. This is performed through the simple bracket method <code>[feature, sample, assay]</code>.</p>
<section id="subset-assays" class="level2">
<h2 class="anchored" data-anchor-id="subset-assays">Subset assays</h2>
<p>Suppose that we want to focus only on the first MS batch (<code>190321S_LCA10_X_FP97AG</code>) for separate processing of the data. Subsetting the <code>QFeatures</code> object for that assay is simply:</p>
<div id="1fc9b772-2ee9-41a3-a25a-de17e9195215" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="10">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>scp1[, , <span class="st">"190321S_LCA10_X_FP97AG"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>harmonizing input:
  removing 103 sampleMap rows not in names(experiments)
  removing 27 colData rownames not in sampleMap 'primary'
</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 1 assays:
 [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns </code></pre>
</div>
</div>
<p>An alternative that results in exactly the same output is using the <code>subsetByAssay</code> method.</p>
<div id="56c6b2e1-dc58-485f-8f0d-4c9094297baa" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="11">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">subsetByAssay</span>(scp1, <span class="st">"190321S_LCA10_X_FP97AG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>harmonizing input:
  removing 103 sampleMap rows not in names(experiments)
  removing 27 colData rownames not in sampleMap 'primary'
</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 1 assays:
 [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns </code></pre>
</div>
</div>
</section>
<section id="subset-samples" class="level2">
<h2 class="anchored" data-anchor-id="subset-samples">Subset samples</h2>
<p>Subsetting samples is often performed after sample QC where we want to keep only quality samples and sample of interest. In our example, the different samples are either technical controls or single-cells (macrophages and monocytes). Suppose we are only interested in macrophages, we can subset the data as follows:</p>
<div id="ee9aa984-1ed8-4140-98fd-33750332230b" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="12">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>scp1[, scp1<span class="sc">$</span>SampleType <span class="sc">==</span> <span class="st">"Macrophage"</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 5 assays:
 [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 7 columns 
 [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 5 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 8 columns 
 [4] peptides: SingleCellExperiment with 539 rows and 20 columns 
 [5] proteins: SingleCellExperiment with 292 rows and 20 columns </code></pre>
</div>
</div>
<p>An alternative that results in exactly the same output is using the <code>subsetByColData</code> method.</p>
<div id="4736224e-d5fb-4862-a991-bf4b4ce1256f" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="13">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">subsetByColData</span>(scp1, scp1<span class="sc">$</span>SampleType <span class="sc">==</span> <span class="st">"Macrophage"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 5 assays:
 [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 7 columns 
 [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 5 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 8 columns 
 [4] peptides: SingleCellExperiment with 539 rows and 20 columns 
 [5] proteins: SingleCellExperiment with 292 rows and 20 columns </code></pre>
</div>
</div>
</section>
<section id="subset-features" class="level2">
<h2 class="anchored" data-anchor-id="subset-features">Subset features</h2>
<p>Subsetting for features does more than simply subsetting for the features of interest, it will also take the features that are linked to that feature. Here is an example, suppose we are interested in the <code>Q02878</code> protein.</p>
<div id="b1416b17-3dc7-4544-bb1f-4eeb63e223e4" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="14">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>scp1[<span class="st">"Q02878"</span>, , ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 5 assays:
 [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 9 rows and 11 columns 
 [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 10 rows and 11 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 0 rows and 16 columns 
 [4] peptides: SingleCellExperiment with 11 rows and 38 columns 
 [5] proteins: SingleCellExperiment with 1 rows and 38 columns </code></pre>
</div>
</div>
<p>You can see it indeed retrieved that protein from the <code>proteins</code> assay, but it also retrieved 11 associated peptides in the <code>peptides</code> assay and 19 associated PSMs in 2 different MS runs.</p>
<p>An alternative that results in exactly the same output is using the <code>subsetByColData</code> method.</p>
<div id="9bf33f4e-e183-4254-9266-265c4845bd96" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="15">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">subsetByFeature</span>(scp1, <span class="st">"Q02878"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 5 assays:
 [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 9 rows and 11 columns 
 [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 10 rows and 11 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 0 rows and 16 columns 
 [4] peptides: SingleCellExperiment with 11 rows and 38 columns 
 [5] proteins: SingleCellExperiment with 1 rows and 38 columns </code></pre>
</div>
</div>
<p>You can also subset features based on the <code>rowData</code>. This is performed by <code>filterFeatures</code>. For example, we want to remove features that are associated to reverse sequence hits.</p>
<div id="9fd6af7b-af0b-4e84-b4b5-176d2b35e6a0" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="16">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">filterFeatures</span>(scp1, <span class="sc">~</span> Reverse <span class="sc">!=</span> <span class="st">"+"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'Reverse' found in 4 out of 5 assay(s)
No filter applied to the following assay(s) because one or more filtering variables are missing in the rowData: proteins.
You can control whether to remove or keep the features using the 'keep' argument (see '?filterFeature').
</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 5 assays:
 [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 126 rows and 11 columns 
 [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 132 rows and 11 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 176 rows and 16 columns 
 [4] peptides: SingleCellExperiment with 422 rows and 38 columns 
 [5] proteins: SingleCellExperiment with 0 rows and 38 columns </code></pre>
</div>
</div>
<p>Note however that if an assay is missing the variable that is used to filter the data (in this case the <code>proteins</code> assay), then all features for that assay are removed.</p>
<p>You can also subset the data based on the feature missingness using <code>filterNA</code>. In this example, we filter out proteins with more than 70 % missing data.</p>
<div id="5e965abb-0d18-4b45-8e1e-8770916d9d79" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="17">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">filterNA</span>(scp1, <span class="at">i =</span> <span class="st">"proteins"</span>, <span class="at">pNA =</span> <span class="fl">0.7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 5 assays:
 [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns 
 [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 11 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 16 columns 
 [4] peptides: SingleCellExperiment with 539 rows and 38 columns 
 [5] proteins: SingleCellExperiment with 105 rows and 38 columns </code></pre>
</div>
</div>
</section>
</section>
<section id="common-processing-steps" class="level1">
<h1>Common processing steps</h1>
<p>We here provide a list of common processing steps that are encountered in single-cell proteomics data processing and that are already available in the <code>QFeatures</code> package.</p>
<p>All functions below require the user to select one or more assays from the <code>QFeatures</code> object. This is passed through the <code>i</code> argument. Note that some datasets may contain hundreds of assays and providing the assay selection manually can become cumbersome. We therefore suggest the user to use regular expression (aka regex) to chose from the <code>names()</code> of the <code>QFeautres</code> object. A detailed cheatsheet about regex in R can be found <a href="https://rstudio-pubs-static.s3.amazonaws.com/74603_76cd14d5983f47408fdf0b323550b846.html">here</a>.</p>
<section id="missing-data-assignment" class="level2">
<h2 class="anchored" data-anchor-id="missing-data-assignment">Missing data assignment</h2>
<p>It often occurs that in MS experiements, 0 values are not true zeros but rather signal that is too weak to be detected. Therefore, it is advised to consider 0 values as missing data (<code>NA</code>). You can use <code>zeroIsNa</code> to automatically convert 0 values to <code>NA</code> in assays of interest. For instance, we here replace missing data in the <code>peptides</code> assay.</p>
<div id="730a7217-dab1-4249-bc77-b361c2ad5206" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="18">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">assay</span>(scp1, <span class="st">"peptides"</span>) <span class="sc">==</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>
FALSE  TRUE 
 5611  1509 </code></pre>
</div>
</div>
<div id="3f8c1836-5e53-44b6-a4bf-940d4d2afe81" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="19">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>scp1 <span class="ot">&lt;-</span><span class="fu">zeroIsNA</span>(scp1, <span class="st">"peptides"</span>)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">assay</span>(scp1, <span class="st">"peptides"</span>) <span class="sc">==</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>
FALSE 
 5611 </code></pre>
</div>
</div>
</section>
<section id="feature-aggregation" class="level2">
<h2 class="anchored" data-anchor-id="feature-aggregation">Feature aggregation</h2>
<p>Shotgun proteomics analyses, bulk as well as single-cell, acquire and quantify peptides. However, biological inference is often performed at protein level. Protein quantitations can be estimated through feature aggregation. This is performed by <code>aggregateFeatures</code>, a function that takes an assay from the <code>Qfeatures</code> object and that aggregates its features with respect to a grouping variable in the <code>rowData</code> (<code>fcol</code>) and an aggregation function.</p>
<div id="9d54a9c9-7b85-49bd-a260-e4e7d2f2ed19" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="20">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">aggregateFeatures</span>(scp1, <span class="at">i =</span> <span class="st">"190321S_LCA10_X_FP97AG"</span>, <span class="at">fcol =</span> <span class="st">"protein"</span>,</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">name =</span> <span class="st">"190321S_LCA10_X_FP97AG_aggr"</span>,</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fun =</span> MsCoreUtils<span class="sc">::</span>robustSummary)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Your row data contain missing values. Please read the relevant
section(s) in the aggregateFeatures manual page regarding the effects
of missing values on data aggregation.

Warning message in rlm.default(X, expression, ...):
“'rlm' failed to converge in 20 steps”</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 6 assays:
 [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns 
 [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 11 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 16 columns 
 [4] peptides: SingleCellExperiment with 539 rows and 38 columns 
 [5] proteins: SingleCellExperiment with 292 rows and 38 columns 
 [6] 190321S_LCA10_X_FP97AG_aggr: SingleCellExperiment with 100 rows and 11 columns </code></pre>
</div>
</div>
<p>You can see that the aggregated function is added as a new assay to the <code>QFeatures</code> object. Note also that, under the hood, <code>aggregateFeatures</code> keeps track of the relationship between the features of the newly aggregated assay and its parent.</p>
</section>
<section id="normalization" class="level2">
<h2 class="anchored" data-anchor-id="normalization">Normalization</h2>
<p>An ubiquituous step that is performed in biological data analysis is normalization that is meant to remove undesired variability and to make different samples comparable. The <code>normalize</code> function offers an interface to a wide variety of normalization methods. See <code>?MsCoreUtils::normalize_matrix</code> for more details about the available normalization methods. Below, we normalize the samples so that they are mean centered.</p>
<div id="ba1ab0bb-3af3-41e3-b84d-a7c0f5bce1cf" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="21">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">normalize</span>(scp1, <span class="st">"proteins"</span>, <span class="at">method =</span> <span class="st">"center.mean"</span>,</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">name =</span> <span class="st">"proteins_mcenter"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 6 assays:
 [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns 
 [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 11 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 16 columns 
 [4] peptides: SingleCellExperiment with 539 rows and 38 columns 
 [5] proteins: SingleCellExperiment with 292 rows and 38 columns 
 [6] proteins_mcenter: SingleCellExperiment with 292 rows and 38 columns </code></pre>
</div>
</div>
<p>Other custom normalization can be applied using the <code>sweep</code> method, where normalization factors have to be supplied manually. As an example, we here normalize the samples using a scaled size factor.</p>
<div id="719b246d-a291-47d7-b05b-efe1d1c2c5d0" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="22">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>sf <span class="ot">&lt;-</span> <span class="fu">colSums</span>(<span class="fu">assay</span>(scp1, <span class="st">"proteins"</span>), <span class="at">na.rm =</span> <span class="cn">TRUE</span>) <span class="sc">/</span> <span class="fl">1E4</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sweep</span>(scp1, <span class="at">i =</span> <span class="st">"proteins"</span>,</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">MARGIN =</span> <span class="dv">2</span>, <span class="do">## 1 = by feature; 2 = by sample</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">STATS =</span> sf, <span class="at">FUN =</span> <span class="st">"/"</span>,</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">name =</span> <span class="st">"proteins_sf"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 6 assays:
 [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns 
 [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 11 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 16 columns 
 [4] peptides: SingleCellExperiment with 539 rows and 38 columns 
 [5] proteins: SingleCellExperiment with 292 rows and 38 columns 
 [6] proteins_sf: SingleCellExperiment with 292 rows and 38 columns </code></pre>
</div>
</div>
</section>
<section id="log-transformation" class="level2">
<h2 class="anchored" data-anchor-id="log-transformation">Log transformation</h2>
<p>The <code>QFeatures</code> package also provide the <code>logTransform</code> function to facilitate the transformation of the quantitative data. We here show its usage by transforming the protein data using a base 2 logarithm with a pseudo-count of one.</p>
<div id="9e45975a-7dba-40c1-904d-1ecc9ca7e5a7" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="23">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">logTransform</span>(scp1, <span class="at">i =</span> <span class="st">"proteins"</span>, <span class="at">base =</span> <span class="dv">2</span>, <span class="at">pc =</span> <span class="dv">1</span>,</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">name =</span> <span class="st">"proteins_log"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 6 assays:
 [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns 
 [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 11 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 16 columns 
 [4] peptides: SingleCellExperiment with 539 rows and 38 columns 
 [5] proteins: SingleCellExperiment with 292 rows and 38 columns 
 [6] proteins_log: SingleCellExperiment with 292 rows and 38 columns </code></pre>
</div>
</div>
</section>
<section id="imputation" class="level2">
<h2 class="anchored" data-anchor-id="imputation">Imputation</h2>
<p>Finally, <code>QFeatures</code> offers an interface to a wide variety of imputation methods to replace missing data by estimated values. The list of available methods is given by <code>?MsCoreUtils::impute_matrix</code>. We demonstrate the use of this function by replacing missing data using KNN imputation.</p>
<div id="5faadcff-a52c-4826-a27a-720c68fff3a4" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="24">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anyNA</span>(<span class="fu">assay</span>(scp1, <span class="st">"proteins"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
TRUE
</div>
</div>
<div id="a160e190-8e1a-4e86-b5d6-93821b53211d" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="25">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>scp1 <span class="ot">&lt;-</span> <span class="fu">impute</span>(scp1, <span class="at">i =</span> <span class="st">"proteins"</span>, <span class="at">method =</span><span class="st">"knn"</span>, <span class="at">k =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required namespace: impute

Imputing along margin 1 (features/rows).

Warning message in knnimp(x, k, maxmiss = rowmax, maxp = maxp):
“284 rows with more than 50 % entries missing;
 mean imputation used for these rows”</code></pre>
</div>
</div>
<div id="e16ab01a-1925-4f81-a07b-2ab2b31741f3" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="26">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anyNA</span>(<span class="fu">assay</span>(scp1, <span class="st">"proteins"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
TRUE
</div>
</div>
</section>
</section>
<section id="data-visualization" class="level1">
<h1>Data visualization</h1>
<p>Visualization of the feature and sample metadata is rather straightforward since those are stored as tables (see section <em>Accessing the data</em>). From those tables, any visualization tool can be applied. Note however that using <code>ggplot2</code> require <code>data.frame</code>s or <code>tibble</code>s but <code>rowData</code> and <code>colData</code> are stored as <code>DFrames</code> objects. You can easily convert one data format to another. For example, we plot the parental ion fraction (measure of spectral purity) for each of the three MS batches.</p>
<div id="cf2f0b5c-971a-43fe-9b77-ff60c79672d8" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="27">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>rd <span class="ot">&lt;-</span> <span class="fu">rbindRowData</span>(scp1, <span class="at">i =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">data.frame</span>(rd)) <span class="sc">+</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">y =</span> PIF,</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">x =</span> assay) <span class="sc">+</span></span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“Removed 64 rows containing non-finite outside the scale range
(`stat_boxplot()`).”</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-03-13-ABC14_files/figure-html/cell-28-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Combining the metadata and the quantitative data is more challenging since the risk of data mismatch is increased. The <code>QFeatures</code> package therefore provides th <code>longFormat</code> function to transform a <code>QFeatures</code> object in a long <code>DFrame</code> table. For instance, we plot the quantitative data distribution for the first assay according to the acquisition channel index and colour with respect to the sample type. Both pieces of information are taken from the <code>colData</code>, so we provide them as <code>colvars</code>.</p>
<div id="9093de06-9776-409f-a3e0-e33657df6f1f" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="28">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>lf <span class="ot">&lt;-</span> <span class="fu">longFormat</span>(scp1[, , <span class="dv">1</span>],</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">colvars =</span> <span class="fu">c</span>(<span class="st">"SampleType"</span>, <span class="st">"Channel"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>harmonizing input:
  removing 141 sampleMap rows not in names(experiments)
  removing 27 colData rownames not in sampleMap 'primary'
</code></pre>
</div>
</div>
<div id="95b136ba-e81f-4bfa-ac4a-edba83445a94" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="29">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">data.frame</span>(lf)) <span class="sc">+</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> Channel,</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">y =</span> value,</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">colour =</span> SampleType) <span class="sc">+</span></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-03-13-ABC14_files/figure-html/cell-30-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>A more in-depth tutorial about data visualization from a <code>QFeatures</code> object is provided in the <code>QFeautres</code> visualization <a href="https://rformassspectrometry.github.io/QFeatures/articles/Visualization.html">vignette</a>.</p>
</section>
<section id="single-cell-proteomics-data-processing-and-analysis" class="level1">
<h1>Single Cell Proteomics data processing and analysis</h1>
<section id="the-scp-package" class="level2">
<h2 class="anchored" data-anchor-id="the-scp-package">The <code>scp</code> package</h2>
<p>The <code>scp</code> package is used to process and analyse mass spectrometry (MS)-based single cell proteomics (SCP) data. The functions rely on a specific data structure that wraps <a href="https://rformassspectrometry.github.io/QFeatures/"><code>QFeatures</code></a> objects (<span class="citation" data-cites="Gatto2023-ry">Gatto and Vanderaa (<a href="#ref-Gatto2023-ry" role="doc-biblioref">2023</a>)</span>) around <a href="http://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html"><code>SingleCellExperiment</code></a> objects (<span class="citation" data-cites="Amezquita2019-bf">Amezquita et al. (<a href="#ref-Amezquita2019-bf" role="doc-biblioref">2020</a>)</span>). This data structure could be seen as Matryoshka dolls were the <code>SingleCellExperiment</code> objects are small dolls contained in the bigger <code>QFeatures</code> doll.</p>
<p>The <code>SingleCellExperiment</code> class provides a dedicated framework for single-cell data. The <code>SingleCellExperiment</code> serves as an interface to many cutting-edge methods for processing, visualizing and analysis single-cell data. More information about the <code>SingleCellExperiment</code> class and associated methods can be found in the <a href="http://bioconductor.org/books/release/OSCA/">OSCA book</a>.</p>
<p>The <code>QFeatures</code> class is a data framework dedicated to manipulate and process MS-based quantitative data. It preserves the relationship between the different levels of information: peptide to spectrum match (PSM) data, peptide data and protein data. The <code>QFeatures</code> package also provides an interface to many utility functions to streamline the processing MS data. More information about MS data analysis tools can be found in the <a href="https://www.rformassspectrometry.org/">RforMassSpectrometry project</a>.</p>
<p>Before running the vignette we need to load the <code>scp</code> package.</p>
<div id="34a116b0-ed21-4681-a16f-fc0db87ec16a" class="cell" data-jupyter="{&quot;outputs_hidden&quot;:true}" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"scp"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We also load <code>ggplot2</code> and <code>dplyr</code> for convenient data manipulation and plotting.</p>
<div id="523daa25-3e30-4a8f-9d19-92c05d69a82f" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"dplyr"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="before-you-start" class="level2">
<h2 class="anchored" data-anchor-id="before-you-start">Before you start</h2>
<p>This vignette will guide you through some common steps of mass spectrometry-based single-cell proteomics (SCP) data analysis. SCP is an emerging field and further research is required to develop a principled analysis workflow. Therefore, we <strong>do not guarantee</strong> that the steps presented here are the best steps for this type of data analysis. This vignette performs the steps that were described in the SCoPE2 landmark paper (<span class="citation" data-cites="Specht2021-jm">Specht et al. (<a href="#ref-Specht2021-jm" role="doc-biblioref">2021</a>)</span>) and that were reproduced in another work using the <code>scp</code> package (<span class="citation" data-cites="Vanderaa2021-ue">Vanderaa and Gatto (<a href="#ref-Vanderaa2021-ue" role="doc-biblioref">2021</a>)</span>). The replication on the full SCoPE2 dataset using <code>scp</code> is available in <a href="https://uclouvain-cbio.github.io/SCP.replication/articles/SCoPE2.html">this vignette</a>. We hope to convince the reader that, although the workflow is probably not optimal, <code>scp</code> has the full potential to perform standardized and principled data analysis. All functions presented here are comprehensively documented, highly modular, can easily be extended with new algorithms. Suggestions, feature requests or bug reports are warmly welcome. Feel free to open an issue in the <a href="https://github.com/UCLouvain-CBIO/scp/issues">GitHub repository</a>.</p>
<p>This workflow can be applied to any MS-based SCP data. The minimal requirement to follow this workflow is that the data should contain the following information:</p>
<ul>
<li><code>runCol</code>/<code>Raw.file</code>: field in the feature data and the sample data that gives the names of MS acquisition runs or files.</li>
<li><code>quantCols</code>: field in the sample data that links to columns in the quantification data and that allows to link samples to MS channels (more details in another <a href="https://uclouvain-cbio.github.io/scp/articles/read_scp.html">vignette</a>).</li>
<li><code>SampleType</code>: field in the sample data that provides the type of sample that is acquired (carrier, reference, single-cell,…). Only needed for multiplexing experiments.</li>
<li><code>Potential.contaminant</code>: field in the feature data that marks contaminant peptides.</li>
<li><code>Reverse</code>: field in the feature data that marks reverse peptides.</li>
<li><code>PIF</code>: field in the feature data that provides spectral purity.</li>
<li><code>PEP</code> or <code>dart_PEP</code>: field in the feature data that provides peptide posterior error probabilities.</li>
<li><code>Modified.sequence</code>: field in the feature data that provides the peptide identifiers.</li>
<li><code>Leading.razor.protein</code>: field in the feature data that provides the protein identifiers.</li>
<li>At least one field in the feature data that contains quantification values. In this case, there are 16 quantification columns named as <code>Reporter.intensity.</code> followed by an index (<code>1</code> to <code>16</code>).</li>
</ul>
<p>Each required field will be described more in detail in the corresponding sections. Names can be adapted by the user to more meaningful ones or adapted to other output tables.</p>
</section>
<section id="read-in-scp-data" class="level2">
<h2 class="anchored" data-anchor-id="read-in-scp-data">Read in SCP data</h2>
<p>The first step is to read in the PSM quantification table generated by, for example, MaxQuant (<span class="citation" data-cites="Tyanova2016-yj">Tyanova, Temu, and Cox (<a href="#ref-Tyanova2016-yj" role="doc-biblioref">2016</a>)</span>). We created a small example data by subsetting the MaxQuant <code>evidence.txt</code> table provided in the SCoPE2 landmark paper (<span class="citation" data-cites="Specht2021-jm">Specht et al. (<a href="#ref-Specht2021-jm" role="doc-biblioref">2021</a>)</span>). The <code>mqScpData</code> table is a typical example of what you would get after reading in a CSV file using <code>read.csv</code> or <code>read.table</code>. See <code>?mqScpData</code> for more information about the table content.</p>
<div id="ebe9ce26-4459-4d00-9c30-8f4361b8327e" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="3">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"mqScpData"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We also provide an example of a sample annotation table that provides useful information about the samples that are present in the example data. See <code>?sampleAnnotation</code> for more information about the table content.</p>
<div id="f2e24c65-8423-46b1-aaba-cce5053a4242" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="4">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"sampleAnnotation"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As a note, the example sample data contains 5 different types of samples (<code>SampleType</code>) that can be found in a TMT-based SCP data set:</p>
<div id="2c53b562-13d8-4376-b7f2-b0ba99203f7b" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="5">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sampleAnnotation<span class="sc">$</span>SampleType)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>
     Blank    Carrier Macrophage   Monocyte  Reference     Unused 
        19          3         20          5          3         14 </code></pre>
</div>
</div>
<ul>
<li>The carrier channels (<code>Carrier</code>) contain 200 cell equivalents and are meant to boost the peptide identification rate.</li>
<li>The normalization channels (<code>Reference</code>) contain 5 cell equivalents and are used to partially correct for between-run variation.</li>
<li>The unused channels (<code>Unused</code>) are channels that are left empty due to isotopic cross-contamination by the carrier channel.</li>
<li>The negative control channels (<code>Blank</code>) contain samples that do not contain any cell but are processed as single-cell samples.</li>
<li>The single-cell sample channels contain the single-cell samples of interest, that are macrophage (<code>Macrophage</code>) or monocyte (<code>Monocyte</code>).</li>
</ul>
<p>Using <code>readSCP</code>, we combine both tables in a <code>QFeatures</code> object formatted as described above.</p>
<div id="cea2ab4d-b250-47f7-baea-ba9a4a25fed8" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="6">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">readSCP</span>(<span class="at">assayData =</span> mqScpData,</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">colData =</span> sampleAnnotation,</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">runCol =</span> <span class="st">"Raw.file"</span>,</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">removeEmptyCols =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Checking arguments.

Loading data as a 'SummarizedExperiment' object.

Splitting data in runs.

Formatting sample annotations (colData).

Formatting data as a 'QFeatures' object.
</code></pre>
</div>
</div>
<div id="9ecf17eb-9889-4b53-9d5a-8bd5e376260b" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="7">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>scp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 4 assays:
 [1] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 395 rows and 11 columns 
 [2] 190321S_LCA10_X_FP97_blank_01: SingleCellExperiment with 109 rows and 11 columns 
 [3] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 487 rows and 11 columns 
 [4] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 370 rows and 16 columns </code></pre>
</div>
</div>
<p>See here that the 3 first assays contain 11 columns that correspond to the TMT-11 labels and the last assay contains 16 columns that correspond to the TMT-16 labels.</p>
<p><strong>Important</strong>: More details about the usage of <code>readSCP()</code> and how to read your own data set are provided in the <code>Load data using readSCP</code> <a href="https://uclouvain-cbio.github.io/scp/articles/read_scp.html">vignette</a>.</p>
<p>Another way to get an overview of the scp object is to plot the <code>QFeatures</code> object. This will create a graph where each node is an assay and links between assays are denoted as edges.</p>
<div id="01f75157-92e4-49dd-97b7-069726f4fd37" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="8">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(scp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-03-13-ABC14_files/figure-html/cell-38-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="quality-control" class="level2">
<h2 class="anchored" data-anchor-id="quality-control">Quality control</h2>
<section id="clean-missing-data" class="level3">
<h3 class="anchored" data-anchor-id="clean-missing-data">Clean missing data</h3>
<p>All single-cell data contain many zeros. The zeros can be biological zeros or technical zeros and differentiating between the two types is not a trivial task. To avoid artefacts in downstream steps, we replace the zeros by the missing value <code>NA</code>. The <code>zeroIsNA</code> function takes the <code>QFeatures</code> object and the name(s) or index/indices of the assay(s) to clean and automatically replaces any zero in the selected quantitative data by <code>NA</code>.</p>
<div id="19dc22e0-ca43-4fbc-82e5-e2897060f897" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="9">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">zeroIsNA</span>(scp, <span class="at">i =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="filter-psms" class="level3">
<h3 class="anchored" data-anchor-id="filter-psms">Filter PSMs</h3>
<p>A common steps in SCP is to filter out low-confidence PSMs. Each PSM assay contains feature meta-information that are stored in the <code>rowData</code> of the assays. The <code>QFeatures</code> package allows to quickly filter the rows of an assay by using these information. The available variables in the <code>rowData</code> are listed below for each assay.</p>
<div id="bb3201e6-aefe-41c8-952a-5b201e3986b5" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="10">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowDataNames</span>(scp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>CharacterList of length 4
[["190222S_LCA9_X_FP94BM"]] uid Sequence Length ... residual participated
[["190321S_LCA10_X_FP97_blank_01"]] uid Sequence ... residual participated
[["190321S_LCA10_X_FP97AG"]] uid Sequence Length ... residual participated
[["190914S_LCB3_X_16plex_Set_21"]] uid Sequence ... residual participated</code></pre>
</div>
</div>
<section id="filter-features-based-on-feature-annotations" class="level4">
<h4 class="anchored" data-anchor-id="filter-features-based-on-feature-annotations">Filter features based on feature annotations</h4>
<p>Below are some examples of criteria that are used to identify low-confidence. The information is readily available since this was computed by MaxQuant:</p>
<ul>
<li>Remove PSMs that are matched to contaminants</li>
<li>Remove PSMs that are matched to the decoy database</li>
<li>Keep PSMs that exhibit a high PIF (parental ion fraction), indicative of the purity of a spectrum</li>
</ul>
<p>We can perform this filtering using the <code>filterFeatures</code> function from <code>QFeatures</code>. <code>filterFeatures</code> automatically accesses the feature annotations and selects the rows that meet the provided condition(s). For instance, <code>Reverse != "+"</code> keeps the rows for which the <code>Reverse</code> variable in the <code>rowData</code> is not <code>"+"</code> (<em>i.e.</em> the PSM is not matched to the decoy database).</p>
<div id="58de23e5-99b3-4b42-913b-a98d914abc7e" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="11">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">filterFeatures</span>(scp,</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>                      <span class="sc">~</span> Reverse <span class="sc">!=</span> <span class="st">"+"</span> <span class="sc">&amp;</span></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>                          Potential.contaminant <span class="sc">!=</span> <span class="st">"+"</span> <span class="sc">&amp;</span></span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>                          <span class="sc">!</span><span class="fu">is.na</span>(PIF) <span class="sc">&amp;</span> PIF <span class="sc">&gt;</span> <span class="fl">0.8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'Reverse' found in 4 out of 4 assay(s)
'Potential.contaminant' found in 4 out of 4 assay(s)
'PIF' found in 4 out of 4 assay(s)

</code></pre>
</div>
</div>
</section>
<section id="filter-assays-based-on-detected-features" class="level4">
<h4 class="anchored" data-anchor-id="filter-assays-based-on-detected-features">Filter assays based on detected features</h4>
<p>To avoid proceeding with failed runs, another interesting filter is to remove assays with too few features. If a batch contains less than, for example, 150 features we can then suspect something wrong happened in that batch and it should be removed. Using <code>dims</code>, we can query the dimensions (hence the number of features and the number of samples) of all assays contained in the dataset.</p>
<div id="3f951330-fc0c-404e-9688-01633033c4fd" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="12">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dims</span>(scp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A matrix: 2 × 4 of type int</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th" scope="col">190222S_LCA9_X_FP94BM</th>
<th data-quarto-table-cell-role="th" scope="col">190321S_LCA10_X_FP97_blank_01</th>
<th data-quarto-table-cell-role="th" scope="col">190321S_LCA10_X_FP97AG</th>
<th data-quarto-table-cell-role="th" scope="col">190914S_LCB3_X_16plex_Set_21</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>283</td>
<td>60</td>
<td>318</td>
<td>200</td>
</tr>
<tr class="even">
<td>11</td>
<td>11</td>
<td>11</td>
<td>16</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Actually, a <code>QFeatures</code> object can be seen as a three-order array: <span class="math inline">\(features \times samples \times assay\)</span>. Hence, <code>QFeatures</code> supports three-order subsetting <code>x[rows, columns, assays]</code>. We first select the assays that have sufficient PSMs (the number of rows is greater than 150), and then subset the <code>scp</code> object for the assays that meet the criterion.</p>
<div id="ce6e26ec-123d-4cc2-8cf6-2c44944dbfb6" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="13">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>keepAssay <span class="ot">&lt;-</span> <span class="fu">dims</span>(scp)[<span class="dv">1</span>, ] <span class="sc">&gt;</span> <span class="dv">150</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> scp[, , keepAssay]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“'experiments' dropped; see 'drops()'”
harmonizing input:
  removing 11 sampleMap rows not in names(experiments)
  removing 11 colData rownames not in sampleMap 'primary'
</code></pre>
</div>
</div>
<div id="9ee84a4e-e9db-449d-844e-e13f39c17c39" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="14">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>scp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 3 assays:
 [1] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 283 rows and 11 columns 
 [2] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 318 rows and 11 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 200 rows and 16 columns </code></pre>
</div>
</div>
<p>Notice the <code>190321S_LCA10_X_FP97_blank_01</code> sample was removed because it did not contain sufficient features, as expected from a blank run. This could also have been the case for failed runs.</p>
</section>
<section id="filter-features-based-on-scp-metrics" class="level4">
<h4 class="anchored" data-anchor-id="filter-features-based-on-scp-metrics">Filter features based on SCP metrics</h4>
<p>Another type of filtering is specific to SCP. In the SCoPE2 analysis, the authors suggest a filters based on the sample to carrier ratio (SCR), that is the reporter ion intensity of a single-cell sample divided by the reporter ion intensity of the carrier channel (200 cells) from the same batch. It is expected that the carrier intensities are much higher than the single-cell intensities.</p>
<p>The SCR can be computed using the <code>computeSCR</code> function from <code>scp</code>. The function must be told which channels are the samples that must be divided and which channel contains the carrier. This information is provided in the sample annotations and is accessed using the <code>colData</code>, under the <code>SampleType</code> field.</p>
<div id="afd61990-8e4b-4900-8f2d-b8dcb84d5f0b" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="15">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">colData</span>(scp)[, <span class="st">"SampleType"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>
     Blank    Carrier Macrophage   Monocyte  Reference     Unused 
         3          3         20          5          3          4 </code></pre>
</div>
</div>
<p>In this dataset, <code>SampleType</code> gives the type of sample that is present in each TMT channel. The SCoPE2 protocole includes 5 types of samples:</p>
<ul>
<li>The carrier channels (<code>Carrier</code>) contain 200 cell equivalents and are meant to boost the peptide identification rate.</li>
<li>The normalization channels (<code>Reference</code>) contain 5 cell equivalents and are used to partially correct for between-run variation.</li>
<li>The unused channels (<code>Unused</code>) are channels that are left empty due to isotopic cross-contamination by the carrier channel.</li>
<li>The negative control channels (<code>Blank</code>) contain samples that do not contain any cell but are processed as single-cell samples.</li>
<li>The single-cell sample channels contain the single-cell samples of interest, that are macrophage (<code>Macrophage</code>) or monocyte (<code>Monocyte</code>).</li>
</ul>
<p>The <code>computeSCR</code> function expects the following input:</p>
<ul>
<li>The <code>QFeatures</code> dataset</li>
<li>The assay name(s) or index/indices for which the SCR should be computed</li>
<li><code>colvar</code>: the variable in the sample annotations (<code>colData</code>) that hold the information used to discriminate sample channels from carrier channels.</li>
<li><code>carrierPattern</code>: a string pattern (following regular expression syntax) that identifies the carrier channel in each batch.</li>
<li><code>samplePattern</code>: a string pattern (following regular expression syntax) that identifies the samples to divide.</li>
</ul>
<p>Optionally, you can also provide the following arguments:</p>
<ul>
<li><code>rowDataName</code>: the name of the column in the <code>rowData</code> where to store the computed SCR for each feature.</li>
<li><code>sampleFUN</code>: when multiple samples are present in an assay, there are as many SCR as there are samples that need to be summarized to a single value per feature. <code>sampleFUN</code> tells which function to use for summarizing the sample values before computing the SCR; the default is the <code>mean</code>.</li>
<li><code>carrierFUN</code>: some designs might include several carriers per run (not the case in this example). Similarly to <code>sampleFUN</code>, <code>carrierFUN</code> tells which function to use for summarizing the carrier values before computing the SCR; the default is the same function as <code>sampleFUN</code>.</li>
</ul>
<p>The function creates a new field in the <code>rowData</code> of the assays. We compute the average SCR for each PSM and store it in the corresponding <code>rowData</code>, under the <code>MeanSCR</code> column.</p>
<div id="f5ec04dc-3cca-4fd3-96cb-e248af5a3f0c" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="16">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">computeSCR</span>(scp,</span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">i =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,</span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">colvar =</span> <span class="st">"SampleType"</span>,</span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">carrierPattern =</span> <span class="st">"Carrier"</span>,</span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">samplePattern =</span> <span class="st">"Macrophage|Monocyte"</span>,</span>
<span id="cb92-6"><a href="#cb92-6" aria-hidden="true" tabindex="-1"></a>                  <span class="at">sampleFUN =</span> <span class="st">"mean"</span>,</span>
<span id="cb92-7"><a href="#cb92-7" aria-hidden="true" tabindex="-1"></a>                  <span class="at">rowDataName =</span> <span class="st">"MeanSCR"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Before applying the filter, we plot the distribution of the average SCR. We collect the <code>rowData</code> from several assays in a single table <code>DataFrame</code> using the <code>rbindRowData</code> function from <code>QFeatures</code>.</p>
<div id="b8b51572-951c-4dad-9f9f-02d8b0c027ed" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="17">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rbindRowData</span>(scp, <span class="at">i =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>) <span class="sc">|&gt;</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> MeanSCR)) <span class="sc">+</span></span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_histogram</span>() <span class="sc">+</span></span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">/</span><span class="dv">200</span>, <span class="fl">0.1</span>),</span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">lty =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb93-7"><a href="#cb93-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
Warning message:
“Removed 15 rows containing non-finite outside the scale range (`stat_bin()`).”</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-03-13-ABC14_files/figure-html/cell-47-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The expected ratio between single cells and the carrier is 1/200 (dashed line). We can see that the distribution mode is slightly shifted towards higher ratios with a mode around 0.01. However, there are a few PSMs that stand out of the distribution and have a much higher signal than expected, indicating something wrong happened during the quantification of those PSMs. We therefore filter out PSMs with an average SCR higher than 0.1 (solide line). This is again easily performed using the <code>filterFeatures</code> functions.</p>
<div id="9ba5cc8b-de55-4d11-9331-1f2127625a74" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="18">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">filterFeatures</span>(scp,</span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>                      <span class="sc">~</span> <span class="sc">!</span><span class="fu">is.na</span>(MeanSCR) <span class="sc">&amp;</span></span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>                          MeanSCR <span class="sc">&lt;</span> <span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'MeanSCR' found in 3 out of 3 assay(s)

</code></pre>
</div>
</div>
</section>
<section id="filter-features-to-control-for-fdr" class="level4">
<h4 class="anchored" data-anchor-id="filter-features-to-control-for-fdr">Filter features to control for FDR</h4>
<p>Finally, we might also want to control for false discovery rate (FDR). MaxQuant already computes posterior error probabilities (PEP), but filtering on PEPs is too conservative (<span class="citation" data-cites="Kall2008-hb">Käll et al. (<a href="#ref-Kall2008-hb" role="doc-biblioref">2008</a>)</span>) so we provide the <code>pep2qvalue</code> function to convert PEPs to q-values that are directly related to FDR. We here compute the q-values from the PEP (<code>dart_PEP</code>) across all 3 assays. <code>dart_PEP</code> contains the PEP values that have been updated using the DART-ID algorithm (<span class="citation" data-cites="Chen2019-uc">Chen, Franks, and Slavov (<a href="#ref-Chen2019-uc" role="doc-biblioref">2019</a>)</span>). The function will store the results in the <code>rowData</code>, we here asked to name the new column <code>qvalue_PSMs</code>.</p>
<div id="d81c8007-24e6-4a0b-afac-bc4c6e49f5a4" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="19">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">pep2qvalue</span>(scp,</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">i =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,</span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">PEP =</span> <span class="st">"dart_PEP"</span>,</span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">rowDataName =</span> <span class="st">"qvalue_PSMs"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We also allow to compute q-values at peptide or protein level rather than PSM. In this case, you need to supply the <code>groupBy</code> argument. Suppose we want to compute the q-values at protein level, we can fetch the protein information stored under <code>Leading.razor.protein</code> in the <code>rowData</code>. This time, we store the q-values in a new field called <code>qvalue_proteins</code>.</p>
<div id="f27a0831-5252-4d3f-bab3-3b8ce75aa9c1" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="20">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">pep2qvalue</span>(scp,</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">i =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">PEP =</span> <span class="st">"dart_PEP"</span>,</span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">groupBy =</span> <span class="st">"Leading.razor.protein"</span>,</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">rowDataName =</span> <span class="st">"qvalue_proteins"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now filter the PSM to control, let’s say, the protein FDR at 1%. This can be performed using <code>filterFeatures</code> because the q-values were stored in the <code>rowData</code>.</p>
<div id="0e2e4aba-0422-47d3-b5ab-d6605ebe8ab2" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="21">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">filterFeatures</span>(scp,</span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>                      <span class="sc">~</span> qvalue_proteins <span class="sc">&lt;</span> <span class="fl">0.01</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'qvalue_proteins' found in 3 out of 3 assay(s)

</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="process-the-psm-data" class="level2">
<h2 class="anchored" data-anchor-id="process-the-psm-data">Process the PSM data</h2>
<section id="relative-reporter-ion-intensity" class="level3">
<h3 class="anchored" data-anchor-id="relative-reporter-ion-intensity">Relative reporter ion intensity</h3>
<p>In order to partialy correct for between-run variation, SCoPE2 suggests computing relative reporter ion intensities. This means that intensities measured for single-cells are divided by the reference channel containing 5-cell equivalents. We use the <code>divideByReference</code> function that divides channels of interest by the reference channel. Similarly to <code>computeSCR</code>, we can point to the samples and the reference columns in each assay using the annotation contained in the <code>colData</code>.</p>
<p>We here divide all columns (using the regular expression wildcard <code>.</code>) by the reference channel (<code>Reference</code>).</p>
<div id="139e65ad-bce8-499f-bb17-5ad6658dcc12" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="22">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">divideByReference</span>(scp,</span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>                         <span class="at">i =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,</span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">colvar =</span> <span class="st">"SampleType"</span>,</span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>                         <span class="at">samplePattern =</span> <span class="st">"."</span>,</span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a>                         <span class="at">refPattern =</span> <span class="st">"Reference"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="aggregate-psm-data-to-peptide-data" class="level2">
<h2 class="anchored" data-anchor-id="aggregate-psm-data-to-peptide-data">Aggregate PSM data to peptide data</h2>
<p>Now that the PSM assays are processed, we can aggregate them to peptides. This is performed using the <code>aggregateFeatures</code> function. For each assay, the function aggregates several PSMs into a unique peptide. This is best illustrated by the figure below.</p>
<div id="ce5ef4ab-f3c2-4e1c-8d7b-3ad31c3176bd" class="cell" data-layout-align="center" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

</div>
</div>
<p>Remember there currently are three assays containing the PSM data.</p>
<div id="c0296b2f-65ba-40e9-af13-a4e27253b644" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="23">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>scp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 3 assays:
 [1] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 239 rows and 11 columns 
 [2] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 277 rows and 11 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 186 rows and 16 columns </code></pre>
</div>
</div>
<p>The PSMs are aggregated over the <code>fcol</code> feature variable, here the modified peptide sequence. We also need to supply an aggregating function that will tell how to combine the quantitative data of the PSMs to aggregate. We here aggregate the PSM data using the median value per sample thanks to the <code>matrixStats:colMedians()</code> function. Other functions can be used and we refer to the documentation <code>?aggregateFeatures</code> for more information about available aggregation functions. The <code>aggregateFeatures()</code> function will create a new assay for each aggregated assay. We name the aggregated assays using the original names and appending <code>peptides_</code> at the start.</p>
<div id="c6332e08-8e59-41bd-baf1-701b3e66c124" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="24">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">aggregateFeatures</span>(scp,</span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>                         <span class="at">i =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,</span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">fcol =</span> <span class="st">"Modified.sequence"</span>,</span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>                         <span class="at">name =</span> <span class="fu">paste0</span>(<span class="st">"peptides_"</span>, <span class="fu">names</span>(scp)),</span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a>                         <span class="at">fun =</span> matrixStats<span class="sc">::</span>colMedians, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Your quantitative and row data contain missing values. Please read the
relevant section(s) in the aggregateFeatures manual page regarding the
effects of missing values on data aggregation.

Your quantitative and row data contain missing values. Please read the
relevant section(s) in the aggregateFeatures manual page regarding the
effects of missing values on data aggregation.

Your quantitative and row data contain missing values. Please read the
relevant section(s) in the aggregateFeatures manual page regarding the
effects of missing values on data aggregation.
</code></pre>
</div>
</div>
<p>Notice that 3 new assays were created in the <code>scp</code> object. Those new assays contain the aggregated features while the three first assays are unchanged. This allows to keep track of the data processing.</p>
<div id="dcccb76a-64eb-4e1c-8243-ff8b4ad7e1c1" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="25">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>scp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 6 assays:
 [1] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 239 rows and 11 columns 
 [2] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 277 rows and 11 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 186 rows and 16 columns 
 [4] peptides_190222S_LCA9_X_FP94BM: SingleCellExperiment with 216 rows and 11 columns 
 [5] peptides_190321S_LCA10_X_FP97AG: SingleCellExperiment with 256 rows and 11 columns 
 [6] peptides_190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 168 rows and 16 columns </code></pre>
</div>
</div>
<p>Under the hood, the <code>QFeatures</code> architecture preserves the relationship between the aggregated assays. See <code>?AssayLinks</code> for more information on relationships between assays. This is illustrated on the <code>QFeatures</code> plot:</p>
<div id="19667002-e8d4-4780-affd-06929388e06f" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="26">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(scp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-03-13-ABC14_files/figure-html/cell-57-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="join-the-scope2-sets-in-one-assay" class="level2">
<h2 class="anchored" data-anchor-id="join-the-scope2-sets-in-one-assay">Join the SCoPE2 sets in one assay</h2>
<p>Up to now, we kept the data belonging to each MS run in separate assays. We now combine all batches into a single assay. This is done using the <code>joinAssays</code> function from the <code>QFeatures</code> package. Note that we now use the aggregated assays, so assay 4 to 6.</p>
<div id="3f53cc0d-e055-477f-9cf9-bdca1253fdb5" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="27">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">joinAssays</span>(scp,</span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">i =</span> <span class="dv">4</span><span class="sc">:</span><span class="dv">6</span>,</span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">name =</span> <span class="st">"peptides"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this case, one new assay is created in the <code>scp</code> object that combines the data from assay 4 to 6. The samples are always distinct so the number of column in the new assay (here <span class="math inline">\(48\)</span>) will always equals the sum of the columns in the assays to join (here <span class="math inline">\(16 + 16 + 16\)</span>). The feature in the joined assay might contain less features than the sum of the rows of the assays to join since common features between assays are joined in a single row.</p>
<div id="670b1421-1c56-4b2c-80de-1e8b128bd048" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="28">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(scp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-03-13-ABC14_files/figure-html/cell-59-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="filter-single-cells" class="level2">
<h2 class="anchored" data-anchor-id="filter-single-cells">Filter single-cells</h2>
<p>Another common step in single-cell data analysis pipelines is to remove low-quality cells. After subsetting for the samples of interest, we will use 2 metrics: the median relative intensities per cell and the median coefficient of variation (CV) per cell.</p>
<section id="filter-samples-of-interest" class="level3">
<h3 class="anchored" data-anchor-id="filter-samples-of-interest">Filter samples of interest</h3>
<p>We can subset the cells of interest, that is the negative control samples, the macrophages and the monocytes. This can easily be done by taking advantage of the <code>colData</code> and the subsetting operators. Recall that <code>QFeatures</code> objects support three-order subsetting, <code>x[rows, columns, assays]</code>, where columns are the samples of interest.</p>
<div id="2b8ee524-1b01-41a3-a6e0-87bc7b2cc99b" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="29">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>scp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 7 assays:
 [1] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 239 rows and 11 columns 
 [2] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 277 rows and 11 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 186 rows and 16 columns 
 [4] peptides_190222S_LCA9_X_FP94BM: SingleCellExperiment with 216 rows and 11 columns 
 [5] peptides_190321S_LCA10_X_FP97AG: SingleCellExperiment with 256 rows and 11 columns 
 [6] peptides_190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 168 rows and 16 columns 
 [7] peptides: SingleCellExperiment with 384 rows and 38 columns </code></pre>
</div>
</div>
<div id="e0645ee7-66a9-4555-aaa0-220670a30731" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="30">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> scp[, scp<span class="sc">$</span>SampleType <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Blank"</span>, <span class="st">"Macrophage"</span>, <span class="st">"Monocyte"</span>), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The subsetting removes unwanted samples from all assays. The filtered data set contains the same number of assays with the same number of features, but the number of columns (hence sampled) decreased.</p>
<div id="0b1188d5-1f7f-4327-87ff-32a0a9de32f4" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="31">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>scp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 7 assays:
 [1] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 239 rows and 8 columns 
 [2] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 277 rows and 8 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 186 rows and 12 columns 
 [4] peptides_190222S_LCA9_X_FP94BM: SingleCellExperiment with 216 rows and 8 columns 
 [5] peptides_190321S_LCA10_X_FP97AG: SingleCellExperiment with 256 rows and 8 columns 
 [6] peptides_190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 168 rows and 12 columns 
 [7] peptides: SingleCellExperiment with 384 rows and 28 columns </code></pre>
</div>
</div>
</section>
<section id="filter-based-on-the-median-relative-intensity" class="level3">
<h3 class="anchored" data-anchor-id="filter-based-on-the-median-relative-intensity">Filter based on the median relative intensity</h3>
<p>We compute the median relative reporter ion intensity for each cell separately and apply a filter based on this statistic. This procedure recalls that of library size filtering commonly performed in scRNA-Seq data analysis, where the library size is the sum of the counts in each single cell. We compute and store the median intensity in the <code>colData</code>.</p>
<div id="63ecb714-a32a-430e-9620-66117333f026" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="32">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>medians <span class="ot">&lt;-</span> <span class="fu">colMedians</span>(<span class="fu">assay</span>(scp[[<span class="st">"peptides"</span>]]), <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a>scp<span class="sc">$</span>MedianRI <span class="ot">&lt;-</span> medians</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Looking at the distribution of the median per cell can highlight low-quality cells.</p>
<div id="b7052657-3aa8-4f86-b5bc-e67579fafcee" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="33">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(scp) <span class="sc">|&gt;</span></span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> MedianRI,</span>
<span id="cb118-5"><a href="#cb118-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">y =</span> SampleType,</span>
<span id="cb118-6"><a href="#cb118-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">fill =</span> SampleType) <span class="sc">+</span></span>
<span id="cb118-7"><a href="#cb118-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb118-8"><a href="#cb118-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-03-13-ABC14_files/figure-html/cell-64-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The negative control samples should not contain any peptide information and are therefore used to assess the amount of background signal. The graph above confirms that the signal measured in single-cells (macrophages and monocytes) is above the background signal, hence no filtering is needed. Would it not be the case, the same procedure as in the previous section can be used for selecting the cells that have an associated median RI lower that a defined threshold.</p>
</section>
<section id="filter-based-on-the-median-cv" class="level3">
<h3 class="anchored" data-anchor-id="filter-based-on-the-median-cv">Filter based on the median CV</h3>
<p>The median CV measures the consistency of quantification for a group of peptides that belong to a protein. We remove cells that exhibit high median CV over the different proteins. We compute the median CV per cell using the <code>computeMedianCV</code> function from the <code>scp</code> package. The function takes the <code>peptides</code> assay and computes the CV for each protein in each cell. To perform this, we must supply the name of the <code>rowData</code> field that contains the protein information through the <code>groupBy</code> argument. We also only want to compute CVs if we have at least 5 peptides per protein. Finally, we also perform a normalization and divide the columns by the median. The computed median CVs are automatically stored in the <code>colData</code> under the name that is supplied, here <code>MedianCV</code>.</p>
<div id="03c4fa5f-c1c9-42fa-aacb-832e5d795632" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="34">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">medianCVperCell</span>(scp,</span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">i =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,</span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">groupBy =</span> <span class="st">"Leading.razor.protein"</span>,</span>
<span id="cb119-4"><a href="#cb119-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">nobs =</span> <span class="dv">5</span>,</span>
<span id="cb119-5"><a href="#cb119-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">norm =</span> <span class="st">"div.median"</span>,</span>
<span id="cb119-6"><a href="#cb119-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">na.rm =</span> <span class="cn">TRUE</span>,</span>
<span id="cb119-7"><a href="#cb119-7" aria-hidden="true" tabindex="-1"></a>                       <span class="at">colDataName =</span> <span class="st">"MedianCV"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The computed CVs are stored in the <code>colData</code> of the <code>peptides</code> assay and holds the median CV per cell computed using at least 5 observations (peptides). The main interest of computing the median CV per cell is to filter cells with reliable quantification. The negative control samples are not expected to have reliable quantifications and hence can be used to estimate an empirical null distribution of the CV. This distribution helps defining a threshold that filters out single-cells that contain noisy quantification.</p>
<div id="b5d7ccb0-1958-408e-a75a-18324bd40479" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="35">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getWithColData</span>(scp, <span class="st">"peptides"</span>) <span class="sc">|&gt;</span></span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colData</span>() <span class="sc">|&gt;</span></span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb120-4"><a href="#cb120-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> MedianCV,</span>
<span id="cb120-5"><a href="#cb120-5" aria-hidden="true" tabindex="-1"></a>               <span class="at">fill =</span> SampleType)) <span class="sc">+</span></span>
<span id="cb120-6"><a href="#cb120-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb120-7"><a href="#cb120-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">0.65</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-03-13-ABC14_files/figure-html/cell-66-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can see that the protein quantification for single-cells are much more consistent than for negative control samples. Based on the distribution of the negative controls, we decide to keep the cells that have a median CV lower than 0.65. Note this example is inaccurate because the null distribution is based on only 3 negative controls, but more sets could lead to a better estimation of the CV null distribution.</p>
<div id="8c329667-0241-49cd-8ef8-baa6af27304b" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="36">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> scp[, <span class="sc">!</span><span class="fu">is.na</span>(scp<span class="sc">$</span>MedianCV) <span class="sc">&amp;</span> scp<span class="sc">$</span>MedianCV <span class="sc">&lt;</span> <span class="fl">0.65</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now remove the negative controls since all QC metrics are now computed.</p>
<div id="3ae8cb5c-a33c-4ff9-8e36-0e665bfe7d72" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="37">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> scp[, scp<span class="sc">$</span>SampleType <span class="sc">!=</span> <span class="st">"Blank"</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="process-the-peptide-data" class="level2">
<h2 class="anchored" data-anchor-id="process-the-peptide-data">Process the peptide data</h2>
<p>In this vignette, the peptide data are further processed before aggregation to proteins. The steps are: normalization, filter peptides based on missing data and log-transformation.</p>
<section id="normalization-1" class="level3">
<h3 class="anchored" data-anchor-id="normalization-1">Normalization</h3>
<p>The columns (samples) of the peptide data are first normalized by dividing the relative intensities by the median relative intensities. Then, the rows (peptides) are normalized by dividing the relative intensities by the mean relative intensities. The normalized data is stored in a separate assay. This normalization procedure is suggested in the SCoPE2 analysis and is applied using the <code>sweep</code> method. Beside the dataset and the assay to normalize, the method expects a <code>MARGIN</code>, that is either row-wise (<code>1</code>) or column-wise (<code>2</code>) transformation, the <code>FUN</code> function to apply and <code>STATS</code>, a vector of values to apply. More conventional normalization procedure can be found in <code>?QFeatures::normalize</code>.</p>
<div id="510b0f80-9b98-47b3-b0db-8248048d3cf1" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="38">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">sweep</span>(scp,</span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">i =</span> <span class="st">"peptides"</span>,</span>
<span id="cb123-3"><a href="#cb123-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">MARGIN =</span> <span class="dv">2</span>,</span>
<span id="cb123-4"><a href="#cb123-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">FUN =</span> <span class="st">"/"</span>,</span>
<span id="cb123-5"><a href="#cb123-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">STATS =</span> <span class="fu">colMedians</span>(<span class="fu">assay</span>(scp[[<span class="st">"peptides"</span>]]), <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb123-6"><a href="#cb123-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">name =</span> <span class="st">"peptides_norm_col"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="9f4d933b-17ac-4134-a070-e8d51c1c779a" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="39">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">sweep</span>(scp,</span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">i =</span> <span class="st">"peptides_norm_col"</span>,</span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">MARGIN =</span> <span class="dv">1</span>,</span>
<span id="cb124-4"><a href="#cb124-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">FUN =</span> <span class="st">"/"</span>,</span>
<span id="cb124-5"><a href="#cb124-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">STATS =</span> <span class="fu">rowMeans</span>(<span class="fu">assay</span>(scp[[<span class="st">"peptides_norm_col"</span>]]),  <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb124-6"><a href="#cb124-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">name =</span> <span class="st">"peptides_norm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Notice each call to <code>sweep</code> created a new assay. Let’s have a look to the current stat of the <code>QFeatures</code> plot:</p>
<div id="4d46d50e-6ffa-49f7-a7bc-9dc379fd72c8" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="40">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(scp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-03-13-ABC14_files/figure-html/cell-71-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="remove-peptides-with-high-missing-rate" class="level3">
<h3 class="anchored" data-anchor-id="remove-peptides-with-high-missing-rate">Remove peptides with high missing rate</h3>
<p>Peptides that contain many missing values are not informative. Therefore, another common procedure is to remove higly missing data. In this example, we remove peptides with more than 99 % missing data. This is done using the <code>filterNA</code> function from <code>QFeatures</code>.</p>
<div id="c6975898-2e14-49a0-86f3-b7e1f3512e95" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="41">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">filterNA</span>(scp,</span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">i =</span> <span class="st">"peptides_norm"</span>,</span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">pNA =</span> <span class="fl">0.99</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="log-transformation-1" class="level3">
<h3 class="anchored" data-anchor-id="log-transformation-1">Log-transformation</h3>
<p>In this vignette, we perform log2-transformation using the <code>logTransform</code> method from <code>QFeatures</code>. Other log-transformation can be applied by changing the <code>base</code> argument.</p>
<div id="0c60dbda-d995-4fc8-9f5c-b4c0ff274d47" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="42">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">logTransform</span>(scp,</span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">base =</span> <span class="dv">2</span>,</span>
<span id="cb127-3"><a href="#cb127-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">i =</span> <span class="st">"peptides_norm"</span>,</span>
<span id="cb127-4"><a href="#cb127-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">name =</span> <span class="st">"peptides_log"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Similarly to <code>sweep</code>, <code>logTransform</code> creates a new assay in <code>scp</code>.</p>
</section>
</section>
<section id="aggregate-peptide-data-to-protein-data" class="level2">
<h2 class="anchored" data-anchor-id="aggregate-peptide-data-to-protein-data">Aggregate peptide data to protein data</h2>
<p>Similarly to aggregating PSM data to peptide data, we can aggregate peptide data to protein data using the <code>aggregateFeatures</code> function.</p>
<div id="a746ec5d-4261-4267-89c1-6953df7712b8" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="43">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">aggregateFeatures</span>(scp,</span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>                         <span class="at">i =</span> <span class="st">"peptides_log"</span>,</span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">name =</span> <span class="st">"proteins"</span>,</span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a>                         <span class="at">fcol =</span> <span class="st">"Leading.razor.protein"</span>,</span>
<span id="cb128-5"><a href="#cb128-5" aria-hidden="true" tabindex="-1"></a>                         <span class="at">fun =</span> matrixStats<span class="sc">::</span>colMedians, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Your quantitative and row data contain missing values. Please read the
relevant section(s) in the aggregateFeatures manual page regarding the
effects of missing values on data aggregation.
</code></pre>
</div>
</div>
<div id="d99d1a77-ca7d-4348-9faa-afbc1de27136" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="44">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>scp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>An instance of class QFeatures containing 11 assays:
 [1] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 239 rows and 6 columns 
 [2] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 277 rows and 6 columns 
 [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 186 rows and 6 columns 
 ...
 [9] peptides_norm: SingleCellExperiment with 373 rows and 18 columns 
 [10] peptides_log: SingleCellExperiment with 373 rows and 18 columns 
 [11] proteins: SingleCellExperiment with 94 rows and 18 columns </code></pre>
</div>
</div>
<p>After the second aggregation, the <code>proteins</code> assay in this example contains quantitative information for 89 proteins in 15 single-cells.</p>
</section>
<section id="process-the-protein-data" class="level2">
<h2 class="anchored" data-anchor-id="process-the-protein-data">Process the protein data</h2>
<p>The protein data is further processed in three steps: normalization, imputation (using the KNN algorithm) and batch correction (using the <code>ComBat</code> algorithm).</p>
<section id="normalization-2" class="level3">
<h3 class="anchored" data-anchor-id="normalization-2">Normalization</h3>
<p>Normalization is performed similarly to peptide normalization. We use the same functions, but since the data were log-transformed at the peptide level, we subtract by the statistic (median or mean) instead of dividing.</p>
<div id="6edd7b9a-410e-46f2-8498-67a365a773c0" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="45">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Center columns with median</span></span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">sweep</span>(scp, <span class="at">i =</span> <span class="st">"proteins"</span>,</span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">MARGIN =</span> <span class="dv">2</span>,</span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">FUN =</span> <span class="st">"-"</span>,</span>
<span id="cb132-5"><a href="#cb132-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">STATS =</span> <span class="fu">colMedians</span>(<span class="fu">assay</span>(scp[[<span class="st">"proteins"</span>]]),</span>
<span id="cb132-6"><a href="#cb132-6" aria-hidden="true" tabindex="-1"></a>                                <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb132-7"><a href="#cb132-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">name =</span> <span class="st">"proteins_norm_col"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="805bffa9-0646-42aa-9f12-b64b3d470f98" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="46">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Center rows with mean</span></span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">sweep</span>(scp, <span class="at">i =</span> <span class="st">"proteins_norm_col"</span>,</span>
<span id="cb133-3"><a href="#cb133-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">MARGIN =</span> <span class="dv">1</span>,</span>
<span id="cb133-4"><a href="#cb133-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">FUN =</span> <span class="st">"-"</span>,</span>
<span id="cb133-5"><a href="#cb133-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">STATS =</span> <span class="fu">rowMeans</span>(<span class="fu">assay</span>(scp[[<span class="st">"proteins_norm_col"</span>]]),</span>
<span id="cb133-6"><a href="#cb133-6" aria-hidden="true" tabindex="-1"></a>                              <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb133-7"><a href="#cb133-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">name =</span> <span class="st">"proteins_norm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="imputation-1" class="level3">
<h3 class="anchored" data-anchor-id="imputation-1">Imputation</h3>
<p>The protein data contains a lot of missing values.</p>
<div id="81706a15-6538-460b-82c6-24953d1231ac" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="47">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a>scp[[<span class="st">"proteins_norm"</span>]] <span class="sc">|&gt;</span></span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">assay</span>() <span class="sc">|&gt;</span></span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">is.na</span>() <span class="sc">|&gt;</span></span>
<span id="cb134-4"><a href="#cb134-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mean</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
0.32919621749409
</div>
</div>
<p>The average missingness in the <code>proteins</code> assay is around 25 %. Including more samples and hence more batches can increase the missingness up to 70 % as seen for the complete SCoPE2 dataset (<span class="citation" data-cites="Specht2021-jm">Specht et al. (<a href="#ref-Specht2021-jm" role="doc-biblioref">2021</a>)</span>). Whether imputation is beneficial or deleterious for the data will not be discussed in this vignette. But taking those missing value into account is essential to avoid artefacts in downstream analyses. The data imputation is performed using the K nearest neighbors algorithm, with k = 3. This is available from the <code>impute()</code> mehtod. More details about the arguments can be found in <code>?impute::impute.knn</code>.</p>
<div id="d41f3b9b-da76-4490-8bb5-0559f8f4c131" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="48">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">impute</span>(scp,</span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">i =</span> <span class="st">"proteins_norm"</span>,</span>
<span id="cb135-3"><a href="#cb135-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">name =</span> <span class="st">"proteins_imptd"</span>,</span>
<span id="cb135-4"><a href="#cb135-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">method =</span> <span class="st">"knn"</span>,</span>
<span id="cb135-5"><a href="#cb135-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">k =</span> <span class="dv">3</span>, <span class="at">rowmax =</span> <span class="dv">1</span>, <span class="at">colmax=</span> <span class="dv">1</span>,</span>
<span id="cb135-6"><a href="#cb135-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">maxp =</span> <span class="cn">Inf</span>, <span class="at">rng.seed =</span> <span class="dv">1234</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required namespace: impute

Imputing along margin 1 (features/rows).
</code></pre>
</div>
</div>
<p>Note that after imputation, no value are missing.</p>
<div id="5c7f1335-f367-4e75-b920-f10d2f1dcbe6" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="49">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>scp[[<span class="st">"proteins_imptd"</span>]] <span class="sc">|&gt;</span></span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">assay</span>() <span class="sc">|&gt;</span></span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">is.na</span>() <span class="sc">|&gt;</span></span>
<span id="cb137-4"><a href="#cb137-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mean</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
0
</div>
</div>
</section>
<section id="batch-correction" class="level3">
<h3 class="anchored" data-anchor-id="batch-correction">Batch correction</h3>
<p>A very important step for processing SCP data is to correct for batch effects. Batch effects are caused by technical variation occurring during different MS runs. Since only a small number of single-cells can be acquired at once, batch effects are unavoidable.</p>
<p>The <code>ComBat()</code> function from the <code>sva</code> package can be used to perform batch correction as it is performed in the SCoPE2 analysis. We do not claim that <code>ComBat</code> is the best algorithm for batch correcting SCP data and other batch correcting methods could be used using the same procedure.</p>
<p>We first extract the assay to process.</p>
<div id="0864acda-7bc1-4a77-b72f-1d013c456d1a" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="50">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">getWithColData</span>(scp, <span class="st">"proteins_imptd"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we need to provide a design matrix and the batch annotation to <code>Combat</code>. The design matrix allows to protect variables of interest, in our case <code>SampleType</code>.</p>
<div id="efa80bb7-d4ce-49ed-abe5-62309ee55930" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="51">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>batch <span class="ot">&lt;-</span> sce<span class="sc">$</span>runCol</span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> SampleType, <span class="at">data =</span> <span class="fu">colData</span>(sce))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We then load and call <code>ComBat</code> and overwrite the data matrix. Recall the data matrix can be accessed using the <code>assay</code> function.</p>
<div id="05a15660-5e7a-46de-90e5-9686af9f357f" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="52">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sva)</span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(sce) <span class="ot">&lt;-</span> <span class="fu">ComBat</span>(<span class="at">dat =</span> <span class="fu">assay</span>(sce),</span>
<span id="cb140-3"><a href="#cb140-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">batch =</span> batch,</span>
<span id="cb140-4"><a href="#cb140-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">mod =</span> model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: mgcv

Loading required package: nlme


Attaching package: ‘nlme’


The following object is masked from ‘package:dplyr’:

    collapse


The following object is masked from ‘package:IRanges’:

    collapse


This is mgcv 1.9-1. For overview type 'help("mgcv-package")'.

Loading required package: genefilter


Attaching package: ‘genefilter’


The following objects are masked from ‘package:MatrixGenerics’:

    rowSds, rowVars


The following objects are masked from ‘package:matrixStats’:

    rowSds, rowVars


Loading required package: BiocParallel

Found3batches

Adjusting for1covariate(s) or covariate level(s)

Standardizing Data across genes

Fitting L/S model and finding priors

Finding parametric adjustments

Adjusting the Data

</code></pre>
</div>
</div>
<p>Finally, we add the batch corrected assay to the <code>QFeatures</code> object and create the feature links.</p>
<div id="8e901349-5740-45e5-973d-a91bbd32e72a" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="53">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">addAssay</span>(scp,</span>
<span id="cb142-2"><a href="#cb142-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">y =</span> sce,</span>
<span id="cb142-3"><a href="#cb142-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">name =</span> <span class="st">"proteins_batchC"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="d92b500d-be27-4067-a237-2dab7f9034e2" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="54">
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a>scp <span class="ot">&lt;-</span> <span class="fu">addAssayLinkOneToOne</span>(scp,</span>
<span id="cb143-2"><a href="#cb143-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">from =</span> <span class="st">"proteins_imptd"</span>,</span>
<span id="cb143-3"><a href="#cb143-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">to =</span> <span class="st">"proteins_batchC"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For the last time, we plot the overview of the fully processed data set:</p>
<div id="d3a7d085-7bef-4c91-9ecc-0c7bdb531d3a" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="55">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(scp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-03-13-ABC14_files/figure-html/cell-86-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="dimension-reduction" class="level2">
<h2 class="anchored" data-anchor-id="dimension-reduction">Dimension reduction</h2>
<p>Because each assay contains <code>SingelCellExperiment</code> objects, we can easily apply methods developed in the scRNA-Seq field. A useful package for dimension reduction on single-cell data is the <code>scater</code>.</p>
<div id="1b57cf40-d0a5-4f5e-8a94-a7931ba9ee89" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="56">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SingleCellExperiment

Loading required package: scuttle
</code></pre>
</div>
</div>
<p>This package provides streamline functions to computes various dimension reduction such as PCA, UMAP, t-SNE, NMF, MDS, ….</p>
<section id="pca" class="level3">
<h3 class="anchored" data-anchor-id="pca">PCA</h3>
<p>PCA can be computed using the <code>runPCA</code> method. It returns a <code>SingleCellExperiment</code> object for which the dimension reduction results are stored in the <code>reducedDim</code> slot.</p>
<div id="875c2037-8dcc-416d-b31d-7badfa63318d" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="57">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a>scp[[<span class="st">"proteins_batchC"</span>]] <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(scp[[<span class="st">"proteins_batchC"</span>]],</span>
<span id="cb147-2"><a href="#cb147-2" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">ncomponents =</span> <span class="dv">5</span>,</span>
<span id="cb147-3"><a href="#cb147-3" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">ntop =</span> <span class="cn">Inf</span>,</span>
<span id="cb147-4"><a href="#cb147-4" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">scale =</span> <span class="cn">TRUE</span>,</span>
<span id="cb147-5"><a href="#cb147-5" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">exprs_values =</span> <span class="dv">1</span>,</span>
<span id="cb147-6"><a href="#cb147-6" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">name =</span> <span class="st">"PCA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The computed PCA can be displayed using the <code>plotReducedDim</code> function. The <code>dimred</code> arguments should give the name of the dimension reduction results to plot, here we called it <code>PCA</code>. The samples are colored by type of sample.</p>
<div id="e82c9f0e-4658-4807-95ee-c0889b29fd41" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="58">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(scp[[<span class="st">"proteins_batchC"</span>]],</span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">dimred =</span> <span class="st">"PCA"</span>,</span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">colour_by =</span> <span class="st">"SampleType"</span>,</span>
<span id="cb148-4"><a href="#cb148-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">point_alpha =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-03-13-ABC14_files/figure-html/cell-89-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>This is a minimalistic example with only a few plotted cells, but the original SCoPE2 dataset contained more than thousand cells.</p>
</section>
<section id="umap" class="level3">
<h3 class="anchored" data-anchor-id="umap">UMAP</h3>
<p>Similarly to PCA, we can compute a UMAP using the <code>runUMAP</code> method. Note however that the UMAP implementation requires a initialization, usually provided by PCA. The previous PCA results are used automatically when supplying <code>dimred = "PCA"</code> (<code>PCA</code> is the name of the dimension reduction result that we supplied in the previous section).</p>
<div id="6942b211-28dc-473f-8635-ce327d8610ab" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="59">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a>scp[[<span class="st">"proteins_batchC"</span>]] <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(scp[[<span class="st">"proteins_batchC"</span>]],</span>
<span id="cb149-2"><a href="#cb149-2" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">ncomponents =</span> <span class="dv">2</span>,</span>
<span id="cb149-3"><a href="#cb149-3" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">ntop =</span> <span class="cn">Inf</span>,</span>
<span id="cb149-4"><a href="#cb149-4" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">scale =</span> <span class="cn">TRUE</span>,</span>
<span id="cb149-5"><a href="#cb149-5" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">exprs_values =</span> <span class="dv">1</span>,</span>
<span id="cb149-6"><a href="#cb149-6" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">n_neighbors =</span> <span class="dv">3</span>,</span>
<span id="cb149-7"><a href="#cb149-7" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">dimred =</span> <span class="st">"PCA"</span>,</span>
<span id="cb149-8"><a href="#cb149-8" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">name =</span> <span class="st">"UMAP"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The computed UMAP can be displayed using the <code>plotReducedDim</code> function. The <code>dimred</code> arguments gives the name of the dimension reduction results to plot, here we called it <code>UMAP</code>. The samples are colored by type of sample.</p>
<div id="ea67dd5b-05a9-4f66-a80d-488e696413fb" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="60">
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(scp[[<span class="st">"proteins_batchC"</span>]],</span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">dimred =</span> <span class="st">"UMAP"</span>,</span>
<span id="cb150-3"><a href="#cb150-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">colour_by =</span> <span class="st">"SampleType"</span>,</span>
<span id="cb150-4"><a href="#cb150-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">point_alpha =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-03-13-ABC14_files/figure-html/cell-91-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The UMAP plot is a very interesting plot for large datasets. A UMAP on this small example dataset is not useful but is shown for illustration.</p>
</section>
</section>
<section id="monitoring-data-processing" class="level2">
<h2 class="anchored" data-anchor-id="monitoring-data-processing">Monitoring data processing</h2>
<p><code>QFeatures</code> keeps the links between the different assays along the processing of the data. This greatly facilitates the visualization of the quantitative data for a features at the different processing levels. For instance, suppose we are interested in the protein <em>Plastin-2</em> (protein ID is <code>P13796</code>). A useful QC is to monitor the data processing at the PSM, peptide and protein level. This can easily be done thanks to the <code>QFeatures</code> framework. Using the <code>subsetByFeature</code>, we can extract the protein of interest and its related features in the other assays. The data is formatted to a long format table that can easily be plugged in the <code>ggplot2</code> visualization tool.</p>
<div id="7a225023-a29c-486c-82c7-2c21a1d2d6f6" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="61">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Get the features related to Plastin-2 (P13796)</span></span>
<span id="cb151-2"><a href="#cb151-2" aria-hidden="true" tabindex="-1"></a><span class="fu">subsetByFeature</span>(scp, <span class="st">"P13796"</span>) <span class="sc">|&gt;</span></span>
<span id="cb151-3"><a href="#cb151-3" aria-hidden="true" tabindex="-1"></a>    <span class="do">## Format the `QFeatures` to a long format table</span></span>
<span id="cb151-4"><a href="#cb151-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">longFormat</span>(<span class="at">colvars =</span> <span class="fu">c</span>(<span class="st">"runCol"</span>, <span class="st">"SampleType"</span>, <span class="st">"quantCols"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb151-5"><a href="#cb151-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb151-6"><a href="#cb151-6" aria-hidden="true" tabindex="-1"></a>    <span class="do">## This is used to preserve ordering of the samples and assays in ggplot2</span></span>
<span id="cb151-7"><a href="#cb151-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">assay =</span> <span class="fu">factor</span>(assay, <span class="at">levels =</span> <span class="fu">names</span>(scp)),</span>
<span id="cb151-8"><a href="#cb151-8" aria-hidden="true" tabindex="-1"></a>           <span class="at">Channel =</span> <span class="fu">sub</span>(<span class="st">"Reporter.intensity."</span>, <span class="st">"TMT-"</span>, quantCols),</span>
<span id="cb151-9"><a href="#cb151-9" aria-hidden="true" tabindex="-1"></a>           <span class="at">Channel =</span> <span class="fu">factor</span>(Channel, <span class="at">levels =</span> <span class="fu">unique</span>(Channel))) <span class="sc">|&gt;</span></span>
<span id="cb151-10"><a href="#cb151-10" aria-hidden="true" tabindex="-1"></a>    <span class="do">## Start plotting</span></span>
<span id="cb151-11"><a href="#cb151-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> Channel, <span class="at">y =</span> value, <span class="at">group =</span> rowname, <span class="at">col =</span> SampleType)) <span class="sc">+</span></span>
<span id="cb151-12"><a href="#cb151-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb151-13"><a href="#cb151-13" aria-hidden="true" tabindex="-1"></a>    <span class="do">## Plot every assay in a separate facet</span></span>
<span id="cb151-14"><a href="#cb151-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="at">facets =</span> <span class="fu">vars</span>(assay), <span class="at">scales =</span> <span class="st">"free_y"</span>, <span class="at">ncol =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb151-15"><a href="#cb151-15" aria-hidden="true" tabindex="-1"></a>    <span class="do">## Annotate plot</span></span>
<span id="cb151-16"><a href="#cb151-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">"Channels"</span>) <span class="sc">+</span></span>
<span id="cb151-17"><a href="#cb151-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Intensity (arbitrary units)"</span>) <span class="sc">+</span></span>
<span id="cb151-18"><a href="#cb151-18" aria-hidden="true" tabindex="-1"></a>    <span class="do">## Improve plot aspect</span></span>
<span id="cb151-19"><a href="#cb151-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>),</span>
<span id="cb151-20"><a href="#cb151-20" aria-hidden="true" tabindex="-1"></a>          <span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="dv">0</span>),</span>
<span id="cb151-21"><a href="#cb151-21" aria-hidden="true" tabindex="-1"></a>          <span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“Removed 838 rows containing missing values or values outside the scale range
(`geom_point()`).”</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-03-13-ABC14_files/figure-html/cell-92-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="session-information" class="level1">
<h1>Session information</h1>
<div id="a4b91c65-f9eb-4ac1-b5ac-7d90ac9ae6fa" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="62">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>R version 4.4.2 (2024-10-31)
Platform: x86_64-conda-linux-gnu
Running under: AlmaLinux 9.4 (Seafoam Ocelot)

Matrix products: default
BLAS/LAPACK: /faststorage/project/HDSSandbox/Workspaces/samuele/ABC14/ABC14-SCProteomics/.pixi/envs/default/lib/libopenblasp-r0.3.29.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.utf-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.utf-8        LC_COLLATE=en_US.utf-8    
 [5] LC_MONETARY=en_US.utf-8    LC_MESSAGES=en_US.utf-8   
 [7] LC_PAPER=en_US.utf-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.utf-8 LC_IDENTIFICATION=C       

time zone: Europe/Copenhagen
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] scater_1.34.0               scuttle_1.16.0             
 [3] SingleCellExperiment_1.28.0 sva_3.54.0                 
 [5] BiocParallel_1.40.0         genefilter_1.88.0          
 [7] mgcv_1.9-1                  nlme_3.1-167               
 [9] dplyr_1.1.4                 ggplot2_3.5.1              
[11] scp_1.16.0                  QFeatures_1.16.0           
[13] MultiAssayExperiment_1.32.0 SummarizedExperiment_1.36.0
[15] Biobase_2.66.0              GenomicRanges_1.58.0       
[17] GenomeInfoDb_1.42.0         IRanges_2.40.0             
[19] S4Vectors_0.44.0            BiocGenerics_0.52.0        
[21] MatrixGenerics_1.18.0       matrixStats_1.5.0          

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      jsonlite_1.9.1          magrittr_2.0.3         
  [4] ggbeeswarm_0.7.2        farver_2.1.2            zlibbioc_1.52.0        
  [7] vctrs_0.6.5             memoise_2.0.1           Cairo_1.6-2            
 [10] base64enc_0.1-3         htmltools_0.5.8.1       S4Arrays_1.6.0         
 [13] BiocBaseUtils_1.8.0     BiocNeighbors_2.0.0     SparseArray_1.6.0      
 [16] plyr_1.8.9              impute_1.80.0           cachem_1.1.0           
 [19] uuid_1.2-1              igraph_2.0.3            lifecycle_1.0.4        
 [22] pkgconfig_2.0.3         rsvd_1.0.5              Matrix_1.7-2           
 [25] R6_2.6.1                fastmap_1.2.0           GenomeInfoDbData_1.2.13
 [28] clue_0.3-66             digest_0.6.37           fdrtool_1.2.18         
 [31] colorspace_2.1-1        AnnotationDbi_1.68.0    irlba_2.3.5.1          
 [34] lpsymphony_1.34.0       RSQLite_2.3.9           beachmat_2.22.0        
 [37] labeling_0.4.3          httr_1.4.7              abind_1.4-5            
 [40] compiler_4.4.2          bit64_4.5.2             withr_3.0.2            
 [43] viridis_0.6.5           DBI_1.2.3               MASS_7.3-64            
 [46] DelayedArray_0.32.0     tools_4.4.2             vipor_0.4.7            
 [49] beeswarm_0.4.0          glue_1.8.0              grid_4.4.2             
 [52] pbdZMQ_0.3-13           cluster_2.1.8           reshape2_1.4.4         
 [55] generics_0.1.3          gtable_0.3.6            tidyr_1.3.1            
 [58] BiocSingular_1.22.0     ScaledMatrix_1.14.0     metapod_1.14.0         
 [61] XVector_0.46.0          ggrepel_0.9.6           pillar_1.10.1          
 [64] stringr_1.5.1           IRdisplay_1.1           limma_3.62.1           
 [67] splines_4.4.2           lattice_0.22-6          FNN_1.1.4.1            
 [70] survival_3.8-3          bit_4.5.0.1             annotate_1.84.0        
 [73] tidyselect_1.2.1        locfit_1.5-9.11         Biostrings_2.74.0      
 [76] gridExtra_2.3           ProtGenerics_1.38.0     edgeR_4.4.0            
 [79] IHW_1.34.0              statmod_1.5.0           stringi_1.8.4          
 [82] UCSC.utils_1.2.0        lazyeval_0.2.2          evaluate_1.0.3         
 [85] codetools_0.2-20        nipals_1.0              MsCoreUtils_1.18.0     
 [88] tibble_3.2.1            cli_3.6.4               uwot_0.2.3             
 [91] IRkernel_1.3.2          xtable_1.8-4            repr_1.1.7             
 [94] munsell_0.5.1           Rcpp_1.0.14             png_0.1-8              
 [97] XML_3.99-0.17           parallel_4.4.2          blob_1.2.4             
[100] AnnotationFilter_1.30.0 viridisLite_0.4.2       slam_0.1-55            
[103] scales_1.3.0            purrr_1.0.4             crayon_1.5.3           
[106] rlang_1.1.5             KEGGREST_1.46.0        </code></pre>
</div>
</div>
</section>
<section id="license" class="level1">
<h1>License</h1>
<p>This vignette is distributed under a <a href="https://creativecommons.org/licenses/by-sa/2.0/">CC BY-SA license</a> license.</p>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Amezquita2019-bf" class="csl-entry" role="listitem">
Amezquita, Robert A, Aaron T L Lun, Etienne Becht, Vince J Carey, Lindsay N Carpp, Ludwig Geistlinger, Federico Marini, et al. 2020. <span>“Orchestrating Single-Cell Analysis with Bioconductor.”</span> <em>Nat. Methods</em> 17 (2): 137–45.
</div>
<div id="ref-Chen2019-uc" class="csl-entry" role="listitem">
Chen, Albert Tian, Alexander Franks, and Nikolai Slavov. 2019. <span>“<span>DART-ID</span> Increases Single-Cell Proteome Coverage.”</span> <em>PLoS Comput. Biol.</em> 15 (7): e1007082.
</div>
<div id="ref-Gatto2023-ry" class="csl-entry" role="listitem">
Gatto, Laurent, and Christophe Vanderaa. 2023. <span>“QFeatures: Quantitative Features for Mass Spectrometry Data.”</span> <a href="https://doi.org/10.18129/B9.bioc.QFeatures">https://doi.org/10.18129/B9.bioc.QFeatures</a>.
</div>
<div id="ref-Kall2008-hb" class="csl-entry" role="listitem">
Käll, Lukas, John D Storey, Michael J MacCoss, and William Stafford Noble. 2008. <span>“Posterior Error Probabilities and False Discovery Rates: Two Sides of the Same Coin.”</span> <em>J. Proteome Res.</em> 7 (1): 40–44.
</div>
<div id="ref-Specht2021-jm" class="csl-entry" role="listitem">
Specht, Harrison, Edward Emmott, Aleksandra A Petelski, R Gray Huffman, David H Perlman, Marco Serra, Peter Kharchenko, Antonius Koller, and Nikolai Slavov. 2021. <span>“Single-Cell Proteomic and Transcriptomic Analysis of Macrophage Heterogeneity Using <span>SCoPE2</span>.”</span> <em>Genome Biol.</em> 22 (1): 50.
</div>
<div id="ref-Tyanova2016-yj" class="csl-entry" role="listitem">
Tyanova, Stefka, Tikira Temu, and Juergen Cox. 2016. <span>“The <span>MaxQuant</span> Computational Platform for Mass Spectrometry-Based Shotgun Proteomics.”</span> <em>Nat. Protoc.</em> 11 (12): 2301–19.
</div>
<div id="ref-Vanderaa2021-ue" class="csl-entry" role="listitem">
Vanderaa, Christophe, and Laurent Gatto. 2021. <span>“Replication of Single-Cell Proteomics Data Reveals Important Computational Challenges.”</span> <em>Expert Review of Proteomics</em> 18 (10): 835–43. <a href="https://doi.org/10.1080/14789450.2021.1988571">https://doi.org/10.1080/14789450.2021.1988571</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    const typesetMath = (el) => {
      if (window.MathJax) {
        // MathJax Typeset
        window.MathJax.typeset([el]);
      } else if (window.katex) {
        // KaTeX Render
        var mathElements = el.getElementsByClassName("math");
        var macros = [];
        for (var i = 0; i < mathElements.length; i++) {
          var texText = mathElements[i].firstChild;
          if (mathElements[i].tagName == "SPAN") {
            window.katex.render(texText.data, mathElements[i], {
              displayMode: mathElements[i].classList.contains('display'),
              throwOnError: false,
              macros: macros,
              fleqn: false
            });
          }
        }
      }
    }
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        typesetMath(container);
        return container.innerHTML
      } else {
        typesetMath(note);
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      typesetMath(note);
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>