<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.515">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Dimitrios Maniatis">
<meta name="dcterms.date" content="2025-01-30">
<meta name="description" content="Pathway enrichment analysis with R package clusterProfiler">

<title>ABC - Accessible Bioinformatics Cafe - ABC.12: Pixi and Desktop on genomeDK, R-clusterProfiler analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-WG01Z2LJ1J"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-WG01Z2LJ1J', { 'anonymize_ip': true});
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
<!-- MailerLite Universal -->
<script>
    (function(w,d,e,u,f,l,n){w[f]=w[f]||function(){(w[f].q=w[f].q||[])
    .push(arguments);},l=d.createElement(e),l.async=1,l.src=u,
    n=d.getElementsByTagName(e)[0],n.parentNode.insertBefore(l,n);})
    (window,document,'script','https://assets.mailerlite.com/js/universal.js','ml');
    ml('account', '1004787');
</script>
<!-- End MailerLite Universal -->

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta property="og:title" content="ABC - Accessible Bioinformatics Cafe - ABC.12: Pixi and Desktop on genomeDK, R-clusterProfiler analysis">
<meta property="og:description" content="Pathway enrichment analysis with R package clusterProfiler">
<meta property="og:image" content="https://abc.au.dk
litesrv._domainkey.birc.au.dk/img/news/clusterProfiler.png">
<meta property="og:site_name" content="ABC - Accessible Bioinformatics Cafe">
<meta property="og:locale" content="en_US">
<meta property="og:image:height" content="600">
<meta property="og:image:width" content="518">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../img/ABClogo.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ABC - Accessible Bioinformatics Cafe</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../news.html"> 
<span class="menu-text">Calendar</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../Documentation.html"> 
<span class="menu-text">Documentation</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-resources" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Resources</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-resources">    
        <li>
    <a class="dropdown-item" href="../binfResources/Databases.html">
 <span class="dropdown-text">Databases</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../binfResources/bash.html">
 <span class="dropdown-text">Bash tips</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../gethelp.html"> 
<span class="menu-text">People&amp;Help</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/AU-ABC"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#slides" id="toc-slides" class="nav-link active" data-scroll-target="#slides">Slides</a></li>
  <li><a href="#tutorial" id="toc-tutorial" class="nav-link" data-scroll-target="#tutorial">Tutorial</a>
  <ul class="collapse">
  <li><a href="#over-representation-analysis-ora" id="toc-over-representation-analysis-ora" class="nav-link" data-scroll-target="#over-representation-analysis-ora">Over Representation Analysis (ORA)</a>
  <ul class="collapse">
  <li><a href="#statistical-test" id="toc-statistical-test" class="nav-link" data-scroll-target="#statistical-test">Statistical Test</a></li>
  <li><a href="#ora-has-some-significant-drawbacks." id="toc-ora-has-some-significant-drawbacks." class="nav-link" data-scroll-target="#ora-has-some-significant-drawbacks.">ORA has some significant drawbacks.</a></li>
  </ul></li>
  <li><a href="#gene-set-enrichment-analysis" id="toc-gene-set-enrichment-analysis" class="nav-link" data-scroll-target="#gene-set-enrichment-analysis">Gene Set Enrichment Analysis</a></li>
  </ul></li>
  <li><a href="#cluster-profiler-examples" id="toc-cluster-profiler-examples" class="nav-link" data-scroll-target="#cluster-profiler-examples">Cluster Profiler examples</a>
  <ul class="collapse">
  <li><a href="#curated-pathway-lists" id="toc-curated-pathway-lists" class="nav-link" data-scroll-target="#curated-pathway-lists">Curated Pathway Lists</a></li>
  <li><a href="#gene-ontology-go" id="toc-gene-ontology-go" class="nav-link" data-scroll-target="#gene-ontology-go">Gene Ontology (GO)</a></li>
  <li><a href="#kegg-pathways" id="toc-kegg-pathways" class="nav-link" data-scroll-target="#kegg-pathways">KEGG Pathways</a></li>
  <li><a href="#other-databases" id="toc-other-databases" class="nav-link" data-scroll-target="#other-databases">Other Databases</a></li>
  <li><a href="#plots" id="toc-plots" class="nav-link" data-scroll-target="#plots">Plots</a>
  <ul class="collapse">
  <li><a href="#converting-gene-lists" id="toc-converting-gene-lists" class="nav-link" data-scroll-target="#converting-gene-lists">Converting Gene Lists</a></li>
  <li><a href="#running-score-plot-interpretation" id="toc-running-score-plot-interpretation" class="nav-link" data-scroll-target="#running-score-plot-interpretation">Running Score Plot Interpretation</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">ABC.12: Pixi and Desktop on genomeDK, R-clusterProfiler analysis</h1>
  <div class="quarto-categories">
    <div class="quarto-category">GenomeDK</div>
    <div class="quarto-category">pixi</div>
    <div class="quarto-category">conda</div>
    <div class="quarto-category">package-environment</div>
    <div class="quarto-category">R</div>
    <div class="quarto-category">clusterProfiler</div>
    <div class="quarto-category">pathway-analysis</div>
  </div>
  </div>

<div>
  <div class="description">
    Pathway enrichment analysis with R package <strong>clusterProfiler</strong>
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Dimitrios Maniatis </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 30, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="slides" class="level1">
<h1>Slides</h1>
<p>Today’s slides with news about GenomeDK and Pixi, and topic presentation about clusterProfiler and pathway analysis.&nbsp;</p>
<p>&nbsp;</p>
<p align="center">
</p><p><a href="https://abc.au.dk/documentation/slides/20250130-ABC12.zip" style="background-color: #4266A1; color: #FFFFFF; padding: 30px 20px; text-decoration: none; border-radius: 5px;"> Download Slides </a></p>
<p></p>
<p>&nbsp;</p>
</section>
<section id="tutorial" class="level1">
<h1>Tutorial</h1>
<p>In this tutorial, we use clusterProfiler for pathway enrichment analysis in R. Pathway enrichment analysis is a method that identifies if a list of genes is over-represented in a pathway. There two widely popular methods for this, Over Reprentation Analysis(ORA) and Gene Set Enrichment Analysis. Both of these can be done by using clusterProfiler and both methods require a list of genes that are the output of a differential expression analysis. We will follow the examples presented by clusterProfiler here: https://yulab-smu.top/biomedical-knowledge-mining-book/enrichment-overview.html</p>
<p>We cover</p>
<ul>
<li><strong>Basic theory</strong>: What are the issues with the test for enrichment, and what does the test means</li>
<li><strong>Analysis</strong>: How to use associated R packages to extrapolate pathway names connected to genes</li>
<li><strong>Result and plots</strong>: Viewing and reading the results correctly with beautiful plots</li>
</ul>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>You should be able to run the tutorial using the packages loaded in the R code below. Many of them are Bioconductor packages.</p>
</div>
</div>
<div id="24a945e4-d3e2-4d94-8d96-08a35470cdb7" class="cell" data-scrolled="true" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(clusterProfiler)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(airway)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(enrichplot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="over-representation-analysis-ora" class="level2">
<h2 class="anchored" data-anchor-id="over-representation-analysis-ora">Over Representation Analysis (ORA)</h2>
<p><strong>ORA</strong> answers the question: “Which pathways are overrepresented in my list of significant genes?”</p>
<p>To answer this, we need a list of <strong>background genes</strong> (also referred to as the <strong>universe</strong>). This is important because overrepresentation is always relative, i.e.&nbsp;overrepresented <strong>compared to what?</strong> Therefore, we essentially create a <strong>contingency table</strong> to determine the significance of the pathway’s representation in the gene list.e. For example:</p>
<div id="f2dc45fe-8151-4bd0-8446-f238bca7d19d" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>mytable <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">not_interesting_genes=</span><span class="fu">c</span>(<span class="dv">2613</span>, <span class="dv">15310</span>), <span class="at">interesting_genes=</span><span class="fu">c</span>(<span class="dv">28</span>, <span class="dv">29</span>))</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(mytable) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"In pathway"</span>, <span class="st">"not in pathway"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>mytable</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 2 × 2</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">not_interesting_genes</th>
<th data-quarto-table-cell-role="th" scope="col">interesting_genes</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">In pathway</td>
<td>2613</td>
<td>28</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">not in pathway</td>
<td>15310</td>
<td>29</td>
</tr>
</tbody>
</table>
</div>
</div>
<section id="statistical-test" class="level3">
<h3 class="anchored" data-anchor-id="statistical-test">Statistical Test</h3>
<p>To determine over-representation, we use <strong>Fisher’s Exact Test</strong>, which is a statistical test commonly applied in contingency table analysis. The test is represented by the following formula:</p>
<p><span class="math display">\[ p = 1 - \sum_{i=0}^{k-1} \frac{\binom{M}{i} \binom{N-M}{n-i}}{\binom{N}{n}} \]</span></p>
<p>Where: - ( M ) is the total number of genes in the pathway, - ( N ) is the total number of genes in the universe, - ( n ) is the number of significant genes (those of interest), - ( k ) is the number of significant genes in the pathway.</p>
<div id="2637e27e-a2da-417e-b161-320bf70fa23e" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fisher.test</span>(mytable, <span class="at">alternative =</span> <span class="st">"greater"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>
    Fisher's Exact Test for Count Data

data:  mytable
p-value = 1
alternative hypothesis: true odds ratio is greater than 1
95 percent confidence interval:
 0.110242      Inf
sample estimates:
odds ratio 
 0.1767937 </code></pre>
</div>
</div>
<p>Since we are testing multiple pathways we should correct for multiple testing.</p>
<div id="ebb3a71d-ee22-4008-afc8-ebcf4dc5d8d8" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>gene_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">Gene_Name =</span> <span class="fu">paste0</span>(<span class="st">"Gene"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>),</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">LogFC =</span> <span class="fu">c</span>(<span class="fl">2.5</span>, <span class="sc">-</span><span class="fl">1.2</span>, <span class="fl">0.8</span>, <span class="sc">-</span><span class="fl">2.3</span>, <span class="fl">3.1</span>, <span class="sc">-</span><span class="fl">0.5</span>, <span class="fl">1.8</span>, <span class="sc">-</span><span class="fl">1.0</span>, <span class="fl">2.0</span>, <span class="sc">-</span><span class="fl">3.0</span>),</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">P_value =</span> <span class="fu">c</span>(<span class="fl">0.001</span>, <span class="fl">0.04</span>, <span class="fl">0.2</span>, <span class="fl">0.005</span>, <span class="fl">0.0001</span>, <span class="fl">0.3</span>, <span class="fl">0.02</span>, <span class="fl">0.06</span>, <span class="fl">0.0005</span>, <span class="fl">0.00001</span>),</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Regulation =</span> <span class="fu">c</span>(<span class="st">"UP"</span>, <span class="st">"DOWN"</span>, <span class="st">"UP"</span>, <span class="st">"DOWN"</span>, <span class="st">"UP"</span>, <span class="st">"DOWN"</span>, <span class="st">"UP"</span>, <span class="st">"DOWN"</span>, <span class="st">"UP"</span>, <span class="st">"DOWN"</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(gene_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Gene_Name LogFC P_value Regulation
1      Gene1   2.5   1e-03         UP
2      Gene2  -1.2   4e-02       DOWN
3      Gene3   0.8   2e-01         UP
4      Gene4  -2.3   5e-03       DOWN
5      Gene5   3.1   1e-04         UP
6      Gene6  -0.5   3e-01       DOWN
7      Gene7   1.8   2e-02         UP
8      Gene8  -1.0   6e-02       DOWN
9      Gene9   2.0   5e-04         UP
10    Gene10  -3.0   1e-05       DOWN</code></pre>
</div>
</div>
</section>
<section id="ora-has-some-significant-drawbacks." class="level3">
<h3 class="anchored" data-anchor-id="ora-has-some-significant-drawbacks.">ORA has some significant drawbacks.</h3>
<ul>
<li>Mainly it is highly dependent on what we (subjectively) will deem to be interesting genes. Usually that would be genes that are differentially expressed.</li>
<li>Additionally, it is highly dependent on the background genes. If there are lot of non annotated genes in the pathways then the power reduces substantially and also creates very biased results depending on which pathway databases you use.</li>
<li>It does not take into account the combination of the p value and the logfc. What if a gene has a fold change of 5 but a p value of 0.051?</li>
</ul>
</section>
</section>
<section id="gene-set-enrichment-analysis" class="level2">
<h2 class="anchored" data-anchor-id="gene-set-enrichment-analysis">Gene Set Enrichment Analysis</h2>
<p>These issues are addressed by Gene Set Enrichment analysis. <strong>GSEA</strong> answers the question “these changes I see between treatment and control are similar to which known list of changes”. The big difference now is that instead of choosing a list of interesting genes we consider all the genes in the output of our degs analysis and we rank them!</p>
<p>The core idea of Gene Set Enrichment Analysis (GSEA) is to assess whether the genes in a predefined gene set (e.g., a pathway) are enriched at the top or bottom of a ranked gene list.</p>
<ol type="1">
<li><p>Gene Ranking: Genes are ranked by a statistic, such as the log-fold change or p-value, representing their differential expression.</p></li>
<li><p>Enrichment Score (ES): The algorithm traverses the ranked gene list and computes a cumulative running score. When encountering a gene in the pathway, the score is increased proportionally to its rank. For genes not in the pathway, the score is decreased. The enrichment score (ES) is the maximum deviation from zero observed during this traversal.</p></li>
<li><p>Significance Testing: The observed ES is compared to a null distribution of ES values generated by permuting the gene labels. The significance is quantified as a p-value, and adjustments for multiple testing are applied.</p></li>
<li><p>Normalization: The ES is normalized (NES) to account for variations in gene set size, making results comparable across pathways.</p></li>
</ol>
<p>Before we go any further lets quickly get a degs list using the airway package and deseq2</p>
<div id="422cd2e5-70ae-4d6d-be55-74d7b4e5ce9e" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(airway)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>cts  <span class="ot">&lt;-</span> <span class="fu">assay</span>(airway) <span class="co"># The counts matrix</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(airway)) <span class="co"># Our metadata</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>metadata<span class="sc">$</span>condition <span class="ot">&lt;-</span> <span class="fu">factor</span>(metadata<span class="sc">$</span>dex) <span class="co"># factor the dex col in condition</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">countData =</span> cts,</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> metadata,</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> <span class="sc">~</span> condition</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds)</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>resdf <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(res)</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating size factors

estimating dispersions

gene-wise dispersion estimates

mean-dispersion relationship

final dispersion estimates

fitting model and testing
</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>log2 fold change (MLE): condition untrt vs trt 
Wald test p-value: condition untrt vs trt 
DataFrame with 6 rows and 6 columns
                  baseMean log2FoldChange     lfcSE      stat    pvalue
                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ENSG00000000003 708.602170      0.3788470  0.173141  2.188082 0.0286636
ENSG00000000005   0.000000             NA        NA        NA        NA
ENSG00000000419 520.297901     -0.2037604  0.100599 -2.025478 0.0428183
ENSG00000000457 237.163037     -0.0340428  0.126279 -0.269584 0.7874802
ENSG00000000460  57.932633      0.1171786  0.301237  0.388992 0.6972820
ENSG00000000938   0.318098      1.7245505  3.493633  0.493627 0.6215698
                     padj
                &lt;numeric&gt;
ENSG00000000003  0.139308
ENSG00000000005        NA
ENSG00000000419  0.183359
ENSG00000000457  0.930572
ENSG00000000460  0.895441
ENSG00000000938        NA</code></pre>
</div>
</div>
<p>Now we can rank the list. There is not on a correct way to rank the gene lists but we would like the ranking to represent both the significance and the fold change. One suggested formula is:</p>
<p><span class="math display">\[
\text{Ranking Score} = \log_{2}(\text{Fold Change}) \times -\log_{10}(\text{P-value})
\]</span></p>
<p>Additionally we can order them by the stat column of the deseq2 output which is the wald statistic. This creates a list where over expressed and significant genes are at the top, under expressed and significant genes in the bottom and combinations of not significant genes in the middle.</p>
<div id="21898b96-9e9b-4bcb-855a-28de5bc2ea9f" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="6">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>resOrdered <span class="ot">&lt;-</span> res[<span class="fu">order</span>(res<span class="sc">$</span>stat),] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="cluster-profiler-examples" class="level1">
<h1>Cluster Profiler examples</h1>
<p><strong>clusterProfiler</strong> makes it easy to perform both ORA and GSEA analyses. It uses two main functions in the background: * <strong><code>enricher()</code></strong> (for ORA) * <strong><code>GSEA()</code></strong> (for GSEA)</p>
<p>These two allow you to submit your own pathways via the TERM2GENE parameter. For example</p>
<div id="47ece741-7522-4837-ac03-0ee8bfcf9993" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="7">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>TERM2GENE <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">TERM =</span> <span class="fu">c</span>(<span class="st">"GO:0005737"</span>, <span class="st">"GO:0008150"</span>, <span class="st">"KEGG:04110"</span>),</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">GENE =</span> <span class="fu">c</span>(<span class="st">"gene1, gene2, gene3"</span>, <span class="st">"gene4, gene5, gene6"</span>, <span class="st">"gene7, gene8"</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Example</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co">#enricher(gene_list, TERM2GENE = TERM2GENE)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="curated-pathway-lists" class="level3">
<h3 class="anchored" data-anchor-id="curated-pathway-lists">Curated Pathway Lists</h3>
<p>ClusterProfiler also has built-in support for curated lists of pathways, such as <strong>KEGG</strong> and <strong>GO</strong>. These are already implemented in the following functions:</p>
<ol type="1">
<li><strong><code>enrichKEGG</code></strong>: Enrichment analysis for <strong>KEGG</strong> pathways.</li>
<li><strong><code>enrichGO</code></strong>: Enrichment analysis for <strong>Gene Ontology (GO)</strong> terms.</li>
<li><strong><code>gseGO()</code></strong>: GSEA for <strong>Gene Ontology (GO)</strong> terms.</li>
<li><strong><code>gseKEGG</code></strong>: GSEA for <strong>KEGG</strong> pathways.</li>
</ol>
<p>Now lets see some examples for GO. For ORA we will use the examples from the guide</p>
<div id="17fafcf8-2573-48b7-be35-11dac57bb47f" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="8">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(geneList, <span class="at">package=</span><span class="st">"DOSE"</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>gene <span class="ot">&lt;-</span> <span class="fu">names</span>(geneList)[<span class="fu">abs</span>(geneList) <span class="sc">&gt;</span> <span class="dv">2</span>] <span class="co"># Create our list of significant genes</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>ego <span class="ot">&lt;-</span> <span class="fu">enrichGO</span>(<span class="at">gene          =</span> gene, <span class="co"># Our genelist</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">universe      =</span> <span class="fu">names</span>(geneList), <span class="co"># The background universe we talked about</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">OrgDb         =</span> org.Hs.eg.db, <span class="co"># Our database </span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">ont           =</span> <span class="st">"CC"</span>, <span class="co"># The Gene Ontology</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span>, <span class="co"># Correction for multiple testint</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">pvalueCutoff  =</span> <span class="fl">0.01</span>, <span class="co"># Our cutoffs</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">qvalueCutoff  =</span> <span class="fl">0.05</span>,</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">readable      =</span> <span class="cn">TRUE</span>) <span class="co"># SetReadable converts the genes to SYMBOL automatically. Very handy.</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ego)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 12</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">ID</th>
<th data-quarto-table-cell-role="th" scope="col">Description</th>
<th data-quarto-table-cell-role="th" scope="col">GeneRatio</th>
<th data-quarto-table-cell-role="th" scope="col">BgRatio</th>
<th data-quarto-table-cell-role="th" scope="col">RichFactor</th>
<th data-quarto-table-cell-role="th" scope="col">FoldEnrichment</th>
<th data-quarto-table-cell-role="th" scope="col">zScore</th>
<th data-quarto-table-cell-role="th" scope="col">pvalue</th>
<th data-quarto-table-cell-role="th" scope="col">p.adjust</th>
<th data-quarto-table-cell-role="th" scope="col">qvalue</th>
<th data-quarto-table-cell-role="th" scope="col">geneID</th>
<th data-quarto-table-cell-role="th" scope="col">Count</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">GO:0000775</td>
<td>GO:0000775</td>
<td>chromosome, centromeric region</td>
<td>21/201</td>
<td>199/11919</td>
<td>0.10552764</td>
<td>6.257631</td>
<td>9.795483</td>
<td>1.849874e-11</td>
<td>5.272141e-09</td>
<td>4.829145e-09</td>
<td>CDCA8/CDC20/CENPE/NDC80/TOP2A/HJURP/SKA1/NEK2/CENPM/CENPN/ERCC6L/MAD2L1/KIF18A/CDT1/BIRC5/EZH2/TTK/NCAPG/AURKB/AURKA/CCNB1</td>
<td>21</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">GO:0000779</td>
<td>GO:0000779</td>
<td>condensed chromosome, centromeric region</td>
<td>18/201</td>
<td>148/11919</td>
<td>0.12162162</td>
<td>7.211981</td>
<td>9.959264</td>
<td>5.132324e-11</td>
<td>5.985415e-09</td>
<td>5.482485e-09</td>
<td>CDC20/CENPE/NDC80/HJURP/SKA1/NEK2/CENPM/CENPN/ERCC6L/MAD2L1/KIF18A/CDT1/BIRC5/TTK/NCAPG/AURKB/AURKA/CCNB1</td>
<td>18</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">GO:0072686</td>
<td>GO:0072686</td>
<td>mitotic spindle</td>
<td>18/201</td>
<td>152/11919</td>
<td>0.11842105</td>
<td>7.022192</td>
<td>9.786254</td>
<td>8.053824e-11</td>
<td>5.985415e-09</td>
<td>5.482485e-09</td>
<td>KIF23/CENPE/ASPM/DLGAP5/SKA1/NUSAP1/TPX2/TACC3/CDK1/MAD2L1/KIF18A/KIF11/TRAT1/AURKB/PRC1/KIFC1/KIF18B/AURKA</td>
<td>18</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">GO:0005819</td>
<td>GO:0005819</td>
<td>spindle</td>
<td>26/201</td>
<td>338/11919</td>
<td>0.07692308</td>
<td>4.561424</td>
<td>8.699252</td>
<td>8.806167e-11</td>
<td>5.985415e-09</td>
<td>5.482485e-09</td>
<td>CDCA8/CDC20/KIF23/CENPE/ASPM/DLGAP5/SKA1/NUSAP1/TPX2/TACC3/NEK2/CDK1/MAD2L1/KIF18A/BIRC5/KIF11/TRAT1/TTK/AURKB/PRC1/KIFC1/KIF18B/KIF20A/AURKA/CCNB1/KIF4A</td>
<td>26</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">GO:0000793</td>
<td>GO:0000793</td>
<td>condensed chromosome</td>
<td>21/201</td>
<td>219/11919</td>
<td>0.09589041</td>
<td>5.686158</td>
<td>9.166827</td>
<td>1.142683e-10</td>
<td>5.985415e-09</td>
<td>5.482485e-09</td>
<td>CDC20/CENPE/NDC80/TOP2A/NCAPH/HJURP/SKA1/NEK2/CENPM/CENPN/ERCC6L/MAD2L1/KIF18A/CDT1/BIRC5/TTK/NCAPG/AURKB/CHEK1/AURKA/CCNB1</td>
<td>21</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">GO:0098687</td>
<td>GO:0098687</td>
<td>chromosomal region</td>
<td>25/201</td>
<td>319/11919</td>
<td>0.07836991</td>
<td>4.647218</td>
<td>8.647710</td>
<td>1.406548e-10</td>
<td>5.985415e-09</td>
<td>5.482485e-09</td>
<td>CDCA8/CDC20/CENPE/NDC80/TOP2A/HJURP/SKA1/NEK2/CENPM/RAD51AP1/CENPN/CDK1/ERCC6L/MAD2L1/KIF18A/CDT1/BIRC5/EZH2/TTK/NCAPG/AURKB/CHEK1/AURKA/CCNB1/MCM5</td>
<td>25</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Similarly we can to <strong>GSEA</strong>. Here we set min and max gene set sizes because gene sets that are too small might not have enough data for reliable statistical analysis, while very large gene sets may not provide useful biological insights because they tend to be too broad. Your call.</p>
<div id="001c6e08-5592-472f-81eb-f662314a17c6" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="9">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>ego3 <span class="ot">&lt;-</span> <span class="fu">gseGO</span>(<span class="at">geneList     =</span> geneList, <span class="co"># Here we have all the genes</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">OrgDb        =</span> org.Hs.eg.db, <span class="co"># Our database</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">ont          =</span> <span class="st">"CC"</span>,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">minGSSize    =</span> <span class="dv">100</span>, <span class="co"># This is the minimum gene set size</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">maxGSSize    =</span> <span class="dv">500</span>, <span class="co"># Maximum geneset size</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>,</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">verbose      =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message in fgseaMultilevel(pathways = pathways, stats = stats, minSize = minSize, :
“For some of the pathways the P-values were likely overestimated. For such pathways log2err is set to NA.”
Warning message in fgseaMultilevel(pathways = pathways, stats = stats, minSize = minSize, :
“For some pathways, in reality P-values are less than 1e-10. You can set the `eps` argument to zero for better estimation.”</code></pre>
</div>
</div>
</section>
<section id="gene-ontology-go" class="level3">
<h3 class="anchored" data-anchor-id="gene-ontology-go">Gene Ontology (GO)</h3>
<p>Gene Ontology (GO) categorizes genes into three domains:</p>
<ul>
<li><strong>Biological Process (BP)</strong>: The biological objectives or processes in which the gene products are involved (e.g., cell cycle, metabolic processes).</li>
<li><strong>Molecular Function (MF)</strong>: The molecular activities performed by gene products (e.g., enzyme activity, receptor binding).</li>
<li><strong>Cellular Component (CC)</strong>: The locations within the cell where gene products are found (e.g., nucleus, mitochondria).</li>
</ul>
</section>
<section id="kegg-pathways" class="level3">
<h3 class="anchored" data-anchor-id="kegg-pathways">KEGG Pathways</h3>
<p>KEGG (Kyoto Encyclopedia of Genes and Genomes) provides a comprehensive database of <strong>metabolic</strong> and <strong>signaling pathways</strong> that help identify biological functions and interactions.</p>
</section>
<section id="other-databases" class="level3">
<h3 class="anchored" data-anchor-id="other-databases">Other Databases</h3>
<ul>
<li><strong>MsigDB</strong>: A Molecular Signature Database that is particularly useful for performing GSEA.</li>
<li><strong>Reactome</strong>: A curated pathway database for functional annotation of genes, focusing on human pathways.</li>
</ul>
</section>
<section id="plots" class="level2">
<h2 class="anchored" data-anchor-id="plots">Plots</h2>
<p>One of the great things about clusterProfiler is that you can very easily create plots with the <span style="color: orange;"><code>enrichplot</code></span> library and they are very well integrated with ggplot. Here we will show some of them as there are many. You can see more here: https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html</p>
<section id="converting-gene-lists" class="level3">
<h3 class="anchored" data-anchor-id="converting-gene-lists">Converting Gene Lists</h3>
<p>When working with gene lists, it’s often necessary to convert between different types of gene identifiers. For instance, you may need to convert between <strong>ENTREZID</strong>, <strong>ENSEMBL</strong>, and <strong>SYMBOL</strong> identifiers, which are commonly used in gene annotation databases. The <strong>bitr()</strong> function from the <strong>clusterProfiler</strong> package is a useful tool for this task.</p>
<div id="9a7a2f49-2cbf-47ff-a972-fb45dfdd5548" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="10">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bitr</span>(gene, <span class="at">fromType =</span> <span class="st">"ENTREZID"</span>,</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">toType =</span> <span class="fu">c</span>(<span class="st">"ENSEMBL"</span>, <span class="st">"SYMBOL"</span>),</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">OrgDb =</span> org.Hs.eg.db)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:many mapping between keys and columns

Warning message in bitr(gene, fromType = "ENTREZID", toType = c("ENSEMBL", "SYMBOL"), :
“0.48% of input gene IDs are fail to map...”</code></pre>
</div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 242 × 3</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">ENTREZID</th>
<th data-quarto-table-cell-role="th" scope="col">ENSEMBL</th>
<th data-quarto-table-cell-role="th" scope="col">SYMBOL</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">1</td>
<td>4312</td>
<td>ENSG00000196611</td>
<td>MMP1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">2</td>
<td>8318</td>
<td>ENSG00000093009</td>
<td>CDC45</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">3</td>
<td>10874</td>
<td>ENSG00000109255</td>
<td>NMU</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">4</td>
<td>55143</td>
<td>ENSG00000134690</td>
<td>CDCA8</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">5</td>
<td>55388</td>
<td>ENSG00000065328</td>
<td>MCM10</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">6</td>
<td>991</td>
<td>ENSG00000117399</td>
<td>CDC20</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">7</td>
<td>6280</td>
<td>ENSG00000163220</td>
<td>S100A9</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">8</td>
<td>2305</td>
<td>ENSG00000111206</td>
<td>FOXM1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">9</td>
<td>9493</td>
<td>ENSG00000137807</td>
<td>KIF23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">10</td>
<td>1062</td>
<td>ENSG00000138778</td>
<td>CENPE</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">11</td>
<td>3868</td>
<td>ENSG00000186832</td>
<td>KRT16</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">12</td>
<td>4605</td>
<td>ENSG00000101057</td>
<td>MYBL2</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">13</td>
<td>9833</td>
<td>ENSG00000165304</td>
<td>MELK</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">14</td>
<td>9133</td>
<td>ENSG00000157456</td>
<td>CCNB2</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">15</td>
<td>6279</td>
<td>ENSG00000143546</td>
<td>S100A8</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">16</td>
<td>10403</td>
<td>ENSG00000080986</td>
<td>NDC80</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">17</td>
<td>8685</td>
<td>ENSG00000019169</td>
<td>MARCO</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">18</td>
<td>597</td>
<td>ENSG00000140379</td>
<td>BCL2A1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">19</td>
<td>7153</td>
<td>ENSG00000131747</td>
<td>TOP2A</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">20</td>
<td>23397</td>
<td>ENSG00000121152</td>
<td>NCAPH</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">21</td>
<td>6278</td>
<td>ENSG00000143556</td>
<td>S100A7</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">22</td>
<td>79733</td>
<td>ENSG00000129173</td>
<td>E2F8</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">23</td>
<td>259266</td>
<td>ENSG00000066279</td>
<td>ASPM</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">24</td>
<td>1381</td>
<td>ENSG00000166426</td>
<td>CRABP1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">25</td>
<td>3627</td>
<td>ENSG00000169245</td>
<td>CXCL10</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">26</td>
<td>27074</td>
<td>ENSG00000078081</td>
<td>LAMP3</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">27</td>
<td>6241</td>
<td>ENSG00000171848</td>
<td>RRM2</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">28</td>
<td>55165</td>
<td>ENSG00000138180</td>
<td>CEP55</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">29</td>
<td>9787</td>
<td>ENSG00000126787</td>
<td>DLGAP5</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">30</td>
<td>7368</td>
<td>ENSG00000174607</td>
<td>UGT8</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">⋮</td>
<td>⋮</td>
<td>⋮</td>
<td>⋮</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">214</td>
<td>64499</td>
<td>ENSG00000197253</td>
<td>TPSB2</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">215</td>
<td>54829</td>
<td>ENSG00000106819</td>
<td>ASPN</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">216</td>
<td>54829</td>
<td>ENSG00000292224</td>
<td>ASPN</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">217</td>
<td>1524</td>
<td>ENSG00000168329</td>
<td>CX3CR1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">218</td>
<td>10234</td>
<td>ENSG00000128606</td>
<td>LRRC17</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">219</td>
<td>1580</td>
<td>ENSG00000142973</td>
<td>CYP4B1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">220</td>
<td>10647</td>
<td>ENSG00000124935</td>
<td>SCGB1D2</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">221</td>
<td>25893</td>
<td>ENSG00000162722</td>
<td>TRIM58</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">222</td>
<td>24141</td>
<td>ENSG00000125869</td>
<td>LAMP5</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">223</td>
<td>10351</td>
<td>ENSG00000141338</td>
<td>ABCA8</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">224</td>
<td>2330</td>
<td>ENSG00000131781</td>
<td>FMO5</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">225</td>
<td>5304</td>
<td>ENSG00000159763</td>
<td>PIP</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">226</td>
<td>79846</td>
<td>ENSG00000105792</td>
<td>CFAP69</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">227</td>
<td>8614</td>
<td>ENSG00000113739</td>
<td>STC2</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">228</td>
<td>2625</td>
<td>ENSG00000107485</td>
<td>GATA3</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">229</td>
<td>7021</td>
<td>ENSG00000008196</td>
<td>TFAP2B</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">230</td>
<td>7802</td>
<td>ENSG00000163879</td>
<td>DNALI1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">231</td>
<td>79689</td>
<td>ENSG00000127954</td>
<td>STEAP4</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">232</td>
<td>11122</td>
<td>ENSG00000196090</td>
<td>PTPRT</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">233</td>
<td>55351</td>
<td>ENSG00000152953</td>
<td>STK32B</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">234</td>
<td>9</td>
<td>ENSG00000171428</td>
<td>NAT1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">235</td>
<td>4239</td>
<td>ENSG00000166482</td>
<td>MFAP4</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">236</td>
<td>5241</td>
<td>ENSG00000082175</td>
<td>PGR</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">237</td>
<td>10551</td>
<td>ENSG00000106541</td>
<td>AGR2</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">238</td>
<td>10974</td>
<td>ENSG00000148671</td>
<td>ADIRF</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">239</td>
<td>79838</td>
<td>ENSG00000103534</td>
<td>TMC5</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">240</td>
<td>79901</td>
<td>ENSG00000071967</td>
<td>CYBRD1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">241</td>
<td>57758</td>
<td>ENSG00000175356</td>
<td>SCUBE2</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">242</td>
<td>4969</td>
<td>ENSG00000106809</td>
<td>OGN</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">243</td>
<td>4969</td>
<td>ENSG00000291350</td>
<td>OGN</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Lets see from our examples:</p>
<div id="a567d9f1-1e3d-4021-a37f-522286c65028" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="11">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>(ego, <span class="at">qscore =</span> <span class="sc">-</span><span class="fu">log</span>(p.adjust, <span class="at">base=</span><span class="dv">10</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">barplot</span>(<span class="at">x=</span><span class="st">"qscore"</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Bar plot from the ORA  above"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-01-30-ABC12_files/figure-html/cell-12-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="73118a4c-1343-45cc-b4dc-a2507e27add1" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="12">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>edox <span class="ot">&lt;-</span> <span class="fu">setReadable</span>(ego, <span class="st">'org.Hs.eg.db'</span>, <span class="st">'ENTREZID'</span>) <span class="co"># Convert to symbol</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">cnetplot</span>(edox, <span class="at">foldChange=</span>geneList, <span class="at">circular =</span> <span class="cn">TRUE</span>, <span class="at">colorEdge =</span> <span class="cn">TRUE</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-01-30-ABC12_files/figure-html/cell-13-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>A very informative plot is a heatplot which you can generate like this:</p>
<div id="f1b17aec-af72-47db-8627-b8f983b8d50a" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="13">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">heatplot</span>(edox, <span class="at">showCategory=</span><span class="dv">5</span>) <span class="co"># Show category dictates show many genesets you see</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">heatplot</span>(edox, <span class="at">foldChange=</span>geneList, <span class="at">showCategory=</span><span class="dv">5</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p1, p2, <span class="at">ncol=</span><span class="dv">1</span>, <span class="at">labels=</span>LETTERS[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-01-30-ABC12_files/figure-html/cell-14-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>One of the most useful ones I find to be the upset plot. For over-representation analysis, upsetplot will calculate the overlaps among different gene sets and for GSEA result, it will plot the fold change distributions of different categories.</p>
<div id="32ea8a28-6aaa-4cfa-9935-cd8c9652b1f7" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="18">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">upsetplot</span>(ego3) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-01-30-ABC12_files/figure-html/cell-15-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="9e11ed29-4fbd-48a5-a8be-b26309397dca" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="26">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.height=</span><span class="dv">20</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ridgeplot</span>(ego3)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.223
</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-01-30-ABC12_files/figure-html/cell-16-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Lastly we have: Running score and preranked list are traditional methods for visualizing the results of Gene Set Enrichment Analysis (GSEA). These visualizations help to understand how a particular gene set is enriched along the ordered gene list, as well as how the enrichment score changes as you progress through the gene list.</p>
<div id="4a66e132-e549-446b-8336-7945df83a77d" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="27">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">gseaplot</span>(ego3, <span class="at">geneSetID =</span> <span class="dv">1</span>, <span class="at">by =</span> <span class="st">"runningScore"</span>, <span class="at">title =</span> ego3<span class="sc">$</span>Description[<span class="dv">1</span>])</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">gseaplot</span>(ego3, <span class="at">geneSetID =</span> <span class="dv">1</span>, <span class="at">by =</span> <span class="st">"preranked"</span>, <span class="at">title =</span> ego3<span class="sc">$</span>Description[<span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="9ac3c674-57a6-4caf-a94b-1064c7e22f27" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="29">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.height=</span><span class="dv">10</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(p1, p2, <span class="at">ncol=</span><span class="dv">1</span>, <span class="at">labels=</span>LETTERS[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="2025-01-30-ABC12_files/figure-html/cell-18-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="running-score-plot-interpretation" class="level3">
<h3 class="anchored" data-anchor-id="running-score-plot-interpretation">Running Score Plot Interpretation</h3>
<p>The running score plot illustrates how the enrichment score (ES) accumulates as you move through the ranked gene list. The key features to observe are:</p>
<ul>
<li>Sharp Peaks: A sharp rise indicates that a cluster of genes from the pathway is highly ranked, contributing strongly to enrichment.</li>
<li>Peak Position: The x-axis location of the peak corresponds to where the pathway’s genes are most enriched within the ranked list.</li>
<li>Running Score Return: If the score drops back down, it suggests that the pathway genes are not uniformly distributed across the gene list.</li>
</ul>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    const typesetMath = (el) => {
      if (window.MathJax) {
        // MathJax Typeset
        window.MathJax.typeset([el]);
      } else if (window.katex) {
        // KaTeX Render
        var mathElements = el.getElementsByClassName("math");
        var macros = [];
        for (var i = 0; i < mathElements.length; i++) {
          var texText = mathElements[i].firstChild;
          if (mathElements[i].tagName == "SPAN") {
            window.katex.render(texText.data, mathElements[i], {
              displayMode: mathElements[i].classList.contains('display'),
              throwOnError: false,
              macros: macros,
              fleqn: false
            });
          }
        }
      }
    }
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        typesetMath(container);
        return container.innerHTML
      } else {
        typesetMath(note);
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      typesetMath(note);
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>